--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -53,7 +_,7 @@
         } catch (Exception var3) {
         }
 
-        final JFrame frame = new JFrame("Purpur Minecraft server"); // Purpur - Improve GUI
+        final JFrame frame = new JFrame("ShreddedPaper Minecraft server"); // Purpur - Improve GUI // ShreddedPaper
         final MinecraftServerGui gui = new MinecraftServerGui(server);
         frame.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
         frame.add(gui);
@@ -61,7 +_,7 @@
         frame.setLocationRelativeTo(null);
         frame.setVisible(true);
         // Paper start - Improve ServerGUI
-        frame.setName("Purpur Minecraft server"); // Purpur - Improve GUI
+        frame.setName("ShreddedPaper Minecraft server"); // Purpur - Improve GUI // ShreddedPaper
         try {
             frame.setIconImage(javax.imageio.ImageIO.read(java.util.Objects.requireNonNull(MinecraftServerGui.class.getClassLoader().getResourceAsStream("logo.png"))));
         } catch (java.io.IOException ignore) {
@@ -71,7 +_,7 @@
             @Override
             public void windowClosing(final WindowEvent event) {
                 if (!gui.isClosing.getAndSet(true)) {
-                    frame.setTitle("Purpur Minecraft server - shutting down!"); // Purpur - Improve GUI
+                    frame.setTitle("ShreddedPaper Minecraft server - shutting down!"); // Purpur - Improve GUI // ShreddedPaper
                     server.halt(true);
                     gui.runFinalizers();
                 }
