--- a/net/minecraft/world/effect/RaidOmenMobEffect.java
+++ b/net/minecraft/world/effect/RaidOmenMobEffect.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.effect;
 
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.ParticleOptions;
 import net.minecraft.server.level.ServerLevel;
@@ -21,7 +_,9 @@
         if (mob instanceof ServerPlayer player && !mob.isSpectator()) {
             BlockPos raidOmenPosition = player.getRaidOmenPosition();
             if (raidOmenPosition != null) {
+                ShreddedPaper.runSync(level, raidOmenPosition, () -> { // ShreddedPaper - run on raidOmenPosition's thread. Do not use ensureSync as createOrExtendRaid can call positions that are outside the player's thread
                 level.getRaids().createOrExtendRaid(player, raidOmenPosition);
+                }); // ShreddedPaper - run on raidOmenPosition's thread
                 player.clearRaidOmenPosition();
                 return false;
             }
