--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -293,7 +_,7 @@
                 // Paper end - Call BlockDispenseEvent
                 List<org.bukkit.craftbukkit.block.CraftBlockState> capturedBlockStates; // Paper
                 org.bukkit.TreeType treeType; // Paper
-                level.captureTreeGeneration = true; // CraftBukkit
+                level.captureTreeGenerationThreadLocal.set(true); // CraftBukkit // ShreddedPaper - use thread local
                 try { // Paper
                 if (!BoneMealItem.growCrop(dispensed, level, target) && !BoneMealItem.growWaterPlant(dispensed, level, target, null)) {
                     this.setSuccess(false);
@@ -302,13 +_,13 @@
                 }
                 // Paper start
                 } finally {
-                    level.captureTreeGeneration = false;
-
-                    capturedBlockStates = new java.util.ArrayList<>(level.capturedBlockStates.values());
-                    level.capturedBlockStates.clear();
-
-                    treeType = net.minecraft.world.level.block.SaplingBlock.treeType;
-                    net.minecraft.world.level.block.SaplingBlock.treeType = null;
+                    level.captureTreeGenerationThreadLocal.remove(); // ShreddedPaper - use thread local
+
+                    capturedBlockStates = new java.util.ArrayList<>(level.capturedBlockStatesThreadLocal.get().values()); // ShreddedPaper - use thread local
+                    level.capturedBlockStatesThreadLocal.get().clear(); // ShreddedPaper - use thread local
+
+                    treeType = net.minecraft.world.level.block.SaplingBlock.treeTypeThreadLocal.get(); // ShreddedPaper - use thread local
+                    net.minecraft.world.level.block.SaplingBlock.treeTypeThreadLocal.remove(); // ShreddedPaper - use thread local
                 }
                 // Paper end
                 // CraftBukkit start
