--- a/net/minecraft/server/level/PlayerSpawnFinder.java
+++ b/net/minecraft/server/level/PlayerSpawnFinder.java
@@ -3,6 +_,8 @@
 import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 import java.util.function.Supplier;
+
+import io.multipaper.shreddedpaper.region.RegionPos;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportedException;
@@ -70,9 +_,9 @@
     private void scheduleNext() {
         int i = this.nextCandidateIndex++;
         if (i < this.candidateCount) {
-            int i1 = (this.offset + this.coprime * i) % this.candidateCount;
-            int i2 = i1 % (this.radius * 2 + 1);
-            int i3 = i1 / (this.radius * 2 + 1);
+            long i1 = ((this.offset + this.coprime * i) % this.candidateCount) * ((this.radius * 2L + 1L) * (this.radius * 2L + 1L) / this.candidateCount); // ShreddedPaper - fix respawn_radius (MC-302446)
+            int i2 = (int) (i1 % (this.radius * 2L + 1L)); // ShreddedPaper - fix respawn_radius (MC-302446)
+            int i3 = (int) (i1 / (this.radius * 2L + 1L)); // ShreddedPaper - fix respawn_radius (MC-302446)
             int i4 = this.spawnSuggestion.getX() + i2 - this.radius;
             int i5 = this.spawnSuggestion.getZ() + i3 - this.radius;
             this.scheduleCandidate(
@@ -146,7 +_,7 @@
                         crashReportCategory.setDetail("Progress", () -> index + " out of " + this.candidateCount);
                         this.finishedFuture.completeExceptionally(new ReportedException(crashReport));
                     }
-                }, this.level.chunkSource.mainThreadProcessor); // Paper - rewrite chunk system
+                }, this.level.getChunkSource().tickingRegions.executorFor(RegionPos.forChunk(sectionPosX, sectionPosZ))); // Paper - rewrite chunk system // ShreddedPaper - run on region
         }
     }
 
