--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -123,6 +_,7 @@
                 MapId id = frame.cachedMapId; // Paper - Perf: Cache map ids on item frames
                 MapItemSavedData data = MapItem.getSavedData(id, this.level);
                 if (data != null) {
+                    synchronized (data) { // ShreddedPaper - thread-safe
                     for (final net.minecraft.server.network.ServerPlayerConnection connection : this.trackedPlayers) { // Paper
                         final ServerPlayer player = connection.getPlayer(); // Paper
                         data.tickCarriedBy(player, itemStack, frame);
@@ -131,6 +_,7 @@
                             player.connection.send(packet);
                         }
                     }
+                    } // ShreddedPaper - thread-safe
                 }
             }
 
@@ -357,7 +_,7 @@
             if (!slots.isEmpty()) {
                 broadcast.accept(new ClientboundSetEquipmentPacket(this.entity.getId(), slots, true)); // Paper - data sanitization
             }
-            ((LivingEntity) this.entity).detectEquipmentUpdates(); // CraftBukkit - SPIGOT-3789: sync again immediately after sending
+            // ((LivingEntity) this.entity).detectEquipmentUpdates(); // CraftBukkit - SPIGOT-3789: sync again immediately after sending // ShreddedPaper - why? - equipment changes should not be run when sending pairing changes
         }
 
         if (!this.entity.getPassengers().isEmpty()) {
