--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -1,8 +_,15 @@
 package net.minecraft.server.level;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import com.google.common.annotations.VisibleForTesting;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
+import io.multipaper.shreddedpaper.config.ShreddedPaperConfiguration;
+import io.multipaper.shreddedpaper.region.LevelChunkRegionMap;
+import io.multipaper.shreddedpaper.region.RegionPos;
+import io.multipaper.shreddedpaper.threading.ShreddedPaperChangesBroadcaster;
+import io.multipaper.shreddedpaper.threading.ShreddedPaperChunkTicker;
+import io.multipaper.shreddedpaper.threading.ShreddedPaperTickThread;
 import it.unimi.dsi.fastutil.longs.LongSet;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet;
@@ -22,6 +_,7 @@
 import net.minecraft.core.SectionPos;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.game.ClientGamePacketListener;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.FileUtil;
 import net.minecraft.util.Util;
 import net.minecraft.util.VisibleForDebug;
@@ -72,7 +_,9 @@
     private final @Nullable ChunkStatus[] lastChunkStatus = new ChunkStatus[4];
     private final @Nullable ChunkAccess[] lastChunk = new ChunkAccess[4];
     private final List<LevelChunk> spawningChunks = new ObjectArrayList<>();
-    private final Set<ChunkHolder> chunkHoldersToBroadcast = new ReferenceOpenHashSet<>();
+    // private final Set<ChunkHolder> chunkHoldersToBroadcast = new ReferenceOpenHashSet<>(); // ShreddedPaper  - handled by the regions
+    private final ShreddedPaperChunkTicker shreddedPaperChunkTicker = new ShreddedPaperChunkTicker(this); // ShreddedPaper
+    public final LevelChunkRegionMap tickingRegions; // ShreddedPaper
     @VisibleForDebug
     private NaturalSpawner.@Nullable SpawnState lastSpawnState;
     // Paper start
@@ -114,7 +_,7 @@
             );
         }
 
-        if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, chunkX, chunkZ)) {
+        if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThread()) { // ShreddedPaper - use managedBlock if this is any tick thread
             ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler.pushChunkWait(this.level, chunkX, chunkZ);
             this.mainThreadProcessor.managedBlock(completable::isDone);
             ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler.popChunkWait();
@@ -198,6 +_,7 @@
         , final SavedDataStorage savedDataStorage // Paper - initialize SavedDataStorage earlier
     ) {
         this.level = level;
+        this.tickingRegions = new LevelChunkRegionMap(this.level); // ShreddedPaper
         this.mainThreadProcessor = new ServerChunkCache.MainThreadExecutor(level);
         this.mainThread = Thread.currentThread();
         Path dataFolder = levelStorage.getDimensionPath(level.dimension()).resolve("data");
@@ -351,14 +_,14 @@
     }
 
     public CompletableFuture<ChunkResult<ChunkAccess>> getChunkFuture(final int x, final int z, final ChunkStatus targetStatus, final boolean loadOrGenerate) {
-        boolean isMainThread = Thread.currentThread() == this.mainThread;
+        boolean isMainThread = TickThread.isTickThreadFor(this.level, x, z); // ShreddedPaper - regions
         CompletableFuture<ChunkResult<ChunkAccess>> serverFuture;
         if (isMainThread) {
             serverFuture = this.getChunkFutureMainThread(x, z, targetStatus, loadOrGenerate);
             this.mainThreadProcessor.managedBlock(serverFuture::isDone);
         } else {
             serverFuture = CompletableFuture.<CompletableFuture<ChunkResult<ChunkAccess>>>supplyAsync(
-                    () -> this.getChunkFutureMainThread(x, z, targetStatus, loadOrGenerate), this.mainThreadProcessor
+                    () -> this.getChunkFutureMainThread(x, z, targetStatus, loadOrGenerate), this.tickingRegions.executorFor(RegionPos.forChunk(x, z)) // ShreddedPaper - regions
                 )
                 .thenCompose(chunk -> (CompletionStage<ChunkResult<ChunkAccess>>)chunk);
         }
@@ -433,6 +_,7 @@
     }
 
     public boolean pollTask() {
+        if (ShreddedPaperTickThread.isShreddedPaperTickThread()) return false; // ShreddedPaper - don't run main thread tasks on worker threads
         return this.mainThreadProcessor.pollTask();
     }
 
@@ -481,28 +_,37 @@
     // CraftBukkit end
 
     @Override
-    public void tick(final BooleanSupplier haveTime, final boolean tickChunks) {
+    public CompletableFuture<Void> tick(final BooleanSupplier haveTime, final boolean tickChunks) { // ShreddedPaper - run async
+        CompletableFuture<Void> future = CompletableFuture.completedFuture(null); // ShreddedPaper - run async
         ProfilerFiller profiler = Profiler.get();
         profiler.push("purge");
+        CompletableFuture<Void> purgeStaleTicketsFuture = CompletableFuture.completedFuture(null); // ShreddedPaper - purgeStaleTickets async
         if (this.level.tickRateManager().runsNormally() || !tickChunks || this.level.spigotConfig.unloadFrozenChunks) { // Spigot
-            this.ticketStorage.purgeStaleTickets(this.chunkMap);
+            if (this.level.getGameTime() % ShreddedPaperConfiguration.get().optimizations.purgeStaleTicketsFrequency == 0) // ShreddedPaper - purgeStaleTicketsFrequency
+            purgeStaleTicketsFuture = purgeStaleTicketsFuture.thenRunAsync(() -> this.ticketStorage.purgeStaleTickets(this.chunkMap), this.level.levelThread); // ShreddedPaper - purgeStaleTickets async
         }
 
         this.runDistanceManagerUpdates();
         profiler.popPush("chunks");
         if (tickChunks) {
             ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getPlayerChunkLoader().tick(); // Paper - rewrite chunk system
-            this.tickChunks();
-            this.chunkMap.tick();
+            future = future.thenCompose(v -> this.tickChunks()); // ShreddedPaper - run async
+            future = future.thenRun(() -> this.chunkMap.tick()); // ShreddedPaper - run async
         }
 
+        future = CompletableFuture.allOf(future, purgeStaleTicketsFuture); // ShreddedPaper - purgeStaleTickets async
+        future = future.thenRun(() -> { // ShreddedPaper - run async
         profiler.popPush("unload");
         this.chunkMap.tick(haveTime);
         profiler.pop();
         this.clearCache();
+        }); // ShreddedPaper - run async
+
+        return future; // ShreddedPaper - run async
     }
 
-    private void tickChunks() {
+    private CompletableFuture<Void>  tickChunks() { // ShreddedPaper - run async
+        CompletableFuture<Void> future = CompletableFuture.completedFuture(null); // ShreddedPaper - run async
         long time = this.level.getGameTime();
         long timeDiff = time - this.lastInhabitedUpdate;
         this.lastInhabitedUpdate = time;
@@ -511,34 +_,41 @@
             profiler.push("pollingChunks");
             if (this.level.tickRateManager().runsNormally()) {
                 profiler.push("tickingChunks");
-                this.tickChunks(profiler, timeDiff);
+                future = future.thenCompose(v -> this.tickChunks(profiler, timeDiff)); // ShreddedPaper - run async
                 profiler.pop();
-            }
+            } else future = future.thenCompose(v -> this.shreddedPaperChunkTicker.tickChunks(0, null, null)); // ShreddedPaper - still tick chunks anyway
 
-            this.broadcastChangedChunks(profiler);
+            // this.broadcastChangedChunks(profiler); // ShreddedPaper - handled in the region
             profiler.pop();
         }
+
+        return future; // ShreddedPaper - run async
     }
 
     private void broadcastChangedChunks(final ProfilerFiller profiler) {
+        if (true) throw new UnsupportedOperationException("Disabled by ShreddedPaper"); // ShreddedPaper - handled in the region, ensure it stays that way
         profiler.push("broadcast");
 
-        for (ChunkHolder chunkHolder : this.chunkHoldersToBroadcast) {
-            LevelChunk chunk = chunkHolder.getChunkToSend(); // Paper - rewrite chunk system
-            if (chunk != null) {
-                chunkHolder.broadcastChanges(chunk);
-            }
-        }
-
-        this.chunkHoldersToBroadcast.clear();
+        // ShreddedPaper start - handled by the regions
+//        for (ChunkHolder chunkHolder : this.chunkHoldersToBroadcast) {
+//            LevelChunk chunk = chunkHolder.getChunkToSend(); // Paper - rewrite chunk system
+//            if (chunk != null) {
+//                chunkHolder.broadcastChanges(chunk);
+//            }
+//        }
+//
+//        this.chunkHoldersToBroadcast.clear();
+        // ShreddedPaper end - handled by the regions
         profiler.pop();
     }
 
-    private void tickChunks(final ProfilerFiller profiler, final long timeDiff) {
+    private CompletableFuture<Void> tickChunks(final ProfilerFiller profiler, final long timeDiff) { // ShreddedPaper - run async
+        CompletableFuture<Void> future = CompletableFuture.completedFuture(null); // ShreddedPaper - run async
         profiler.push("naturalSpawnCount");
         int chunkCount = this.distanceManager.getNaturalSpawnChunkCount();
         // Paper start - Optional per player mob spawns
         NaturalSpawner.SpawnState spawnCookie;
+        if (ShreddedPaperChunkTicker.willTrySpawnMobsThisTick(this.level)) { // ShreddedPaper - only create NaturalSpawner if we'll use it
         if ((this.spawnFriendlies || this.spawnEnemies) && this.level.paperConfig().entities.spawning.perPlayerMobSpawns) { // don't count mobs when animals and monsters are disabled
             // re-set mob counts
             for (ServerPlayer player : this.level.players()) {
@@ -558,11 +_,13 @@
         } else {
             spawnCookie = NaturalSpawner.createState(chunkCount, this.level.getAllEntities(), this::getFullChunk, !this.level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(this.chunkMap) : null, false);
         }
+        } else { spawnCookie = null; } // ShreddedPaper - only create NaturalSpawner if we'll use it
         // Paper end - Optional per player mob spawns
         this.lastSpawnState = spawnCookie;
         boolean doMobSpawning = this.level.getGameRules().get(GameRules.SPAWN_MOBS) && !this.level.players().isEmpty(); // CraftBukkit
         int tickSpeed = this.level.getGameRules().get(GameRules.RANDOM_TICK_SPEED);
         List<MobCategory> spawningCategories;
+        if (spawnCookie != null) { // ShreddedPaper - only create NaturalSpawner if we'll use it
         if (doMobSpawning && (this.spawnEnemies || this.spawnFriendlies)) { // Paper
             // Paper start - PlayerNaturallySpawnCreaturesEvent
             for (ServerPlayer player : this.level.players()) {
@@ -577,37 +_,42 @@
         } else {
             spawningCategories = List.of();
         }
-
-        List<LevelChunk> spawningChunks = this.spawningChunks;
-
-        try {
-            profiler.popPush("filteringSpawningChunks");
-            this.chunkMap.collectSpawningChunks(spawningChunks);
-            profiler.popPush("shuffleSpawningChunks");
-            // Paper start - chunk tick iteration optimisation
-            this.shuffleRandom.setSeed(this.level.getRandom().nextLong());
-            if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) Util.shuffle(spawningChunks, this.level.getRandom()); // Paper - Optional per player mob spawns; do not need this when per-player is enabled
-            // Paper end - chunk tick iteration optimisation
-            profiler.popPush("tickSpawningChunks");
-
-            for (LevelChunk chunk : spawningChunks) {
-                this.tickSpawningChunk(chunk, timeDiff, spawningCategories, spawnCookie);
-            }
-        } finally {
-            spawningChunks.clear();
-        }
+        } else { spawningCategories = null; } // ShreddedPaper - only create NaturalSpawner if we'll use it
+
+        future = future.thenCompose(v -> this.shreddedPaperChunkTicker.tickChunks(timeDiff, spawningCategories, spawnCookie)).thenRunAsync(() -> {}, this.level.levelThread); // ShreddedPaper - run async
+        // ShreddedPaper start - moved to ShreddedPaperChunkTicker
+//        List<LevelChunk> spawningChunks = this.spawningChunks;
+//
+//        try {
+//            profiler.popPush("filteringSpawningChunks");
+//            this.chunkMap.collectSpawningChunks(spawningChunks);
+//            profiler.popPush("shuffleSpawningChunks");
+//            // Paper start - chunk tick iteration optimisation
+//            this.shuffleRandom.setSeed(this.level.getRandom().nextLong());
+//            if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) Util.shuffle(spawningChunks, this.level.getRandom()); // Paper - Optional per player mob spawns; do not need this when per-player is enabled
+//            // Paper end - chunk tick iteration optimisation
+//            profiler.popPush("tickSpawningChunks");
+//
+//            for (LevelChunk chunk : spawningChunks) {
+//                this.tickSpawningChunk(chunk, timeDiff, spawningCategories, spawnCookie);
+//            }
+//        } finally {
+//            spawningChunks.clear();
+//        }
+        // ShreddedPaper end - moved to ShreddedPaperChunkTicker
 
         profiler.popPush("tickTickingChunks");
-        this.iterateTickingChunksFaster(); // Paper - chunk tick iteration optimizations
+        // this.iterateTickingChunksFaster(); // Paper - chunk tick iteration optimizations // ShreddedPaper - moved to ShreddedPaperChunkTicker
         if (doMobSpawning) {
             profiler.popPush("customSpawners");
-            this.level.tickCustomSpawners(this.spawnEnemies);
+            future = future.thenRun(() -> this.level.tickCustomSpawners(this.spawnEnemies)); // ShreddedPaper - run async
         }
 
         profiler.pop();
+        return future; // ShreddedPaper - run async
     }
 
-    private void tickSpawningChunk(
+    public void tickSpawningChunk( // ShreddedPaper - make public
         final LevelChunk chunk, final long timeDiff, final List<MobCategory> spawningCategories, final NaturalSpawner.SpawnState spawnCookie
     ) {
         ChunkPos chunkPos = chunk.getPos();
@@ -665,16 +_,16 @@
         int zc = SectionPos.blockToSectionCoord(pos.getZ());
         ChunkHolder chunk = this.getVisibleChunkIfPresent(ChunkPos.pack(xc, zc));
         if (chunk != null && chunk.blockChanged(pos)) {
-            this.chunkHoldersToBroadcast.add(chunk);
+            ShreddedPaperChangesBroadcaster.add(chunk); // ShreddedPaper  - handled by the regions
         }
     }
 
     @Override
     public void onLightUpdate(final LightLayer layer, final SectionPos pos) {
-        this.mainThreadProcessor.execute(() -> {
+        this.level.moonrise$getChunkTaskScheduler().scheduleChunkTask(pos.x(), pos.z(), () -> { // ShreddedPaper - run on chunk thread
             ChunkHolder chunk = this.getVisibleChunkIfPresent(pos.chunk().pack());
             if (chunk != null && chunk.sectionLightChanged(layer, pos.y())) {
-                this.chunkHoldersToBroadcast.add(chunk);
+                ShreddedPaperChangesBroadcaster.add(chunk); // ShreddedPaper  - handled by the regions
             }
         });
     }
@@ -790,7 +_,7 @@
 
     public void onChunkReadyToSend(final ChunkHolder chunk) {
         if (chunk.hasChangesToBroadcast()) {
-            this.chunkHoldersToBroadcast.add(chunk);
+            ShreddedPaperChangesBroadcaster.add(chunk); // ShreddedPaper  - handled by the regions
         }
     }
 
@@ -827,13 +_,28 @@
         }
 
         @Override
-        protected boolean pollTask() {
+        public boolean pollTask() { // ShreddedPaper - make public
+            // ShreddedPaper start - run all region tasks cause the worker threads might be deadlocked and not get to them
+            if ((!MinecraftServer.getServer().executingMidTickTasks.get() && TickThread.isTickThread()) || TickThread.isShutdownThread()) {
+                level.getChunkSource().tickingRegions.forEach(region -> {
+                    if (TickThread.isShutdownThread()) {
+                        // We're trying to shut down the server, just execute all the tasks
+                        region.getInternalTaskQueue().executeTask();
+                    } else if (!region.getInternalTaskQueue().hasNoScheduledTasks()) {
+                        // We're the main thread, try executing chunk tasks for regions not being ticked
+                        level.chunkScheduler.getRegionLocker().tryReadOnlyLockNow(region.getRegionPos(), () -> {
+                            region.getInternalTaskQueue().executeTask();
+                        });
+                    }
+                });
+            }
+            // ShreddedPaper end
             // Paper start - rewrite chunk system
             final ServerChunkCache serverChunkCache = ServerChunkCache.this;
             if (serverChunkCache.runDistanceManagerUpdates()) {
                 return true;
             } else {
-                return super.pollTask() | ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)serverChunkCache.level).moonrise$getChunkTaskScheduler().executeMainThreadTask();
+                return (!ShreddedPaperTickThread.isShreddedPaperTickThread() && super.pollTask()) | ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)serverChunkCache.level).moonrise$getChunkTaskScheduler().executeMainThreadTask(); // ShreddedPaper - don't run main thread tasks on workers
             }
             // Paper end - rewrite chunk system
         }
