--- a/net/minecraft/util/thread/BlockableEventLoop.java
+++ b/net/minecraft/util/thread/BlockableEventLoop.java
@@ -179,8 +_,14 @@
             return false;
         }
 
-        this.doRunTask(this.pendingRunnables.remove());
+        // ShreddedPaper start - concurrency fix
+        R task2 = this.pendingRunnables.poll();
+        if (task2 == null) {
+            return false;
+        }
+        this.doRunTask(task2);
         return true;
+        // ShreddedPaper end - concurrency fix
     }
 
     public void managedBlock(final BooleanSupplier condition) {
