--- a/net/minecraft/server/commands/ClearInventoryCommands.java
+++ b/net/minecraft/server/commands/ClearInventoryCommands.java
@@ -6,7 +_,10 @@
 import com.mojang.brigadier.exceptions.DynamicCommandExceptionType;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
 import java.util.function.Predicate;
+
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.commands.CommandBuildContext;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
@@ -65,9 +_,16 @@
         int count = 0;
 
         for (ServerPlayer player : players) {
-            count += player.getInventory().clearOrCountMatchingItems(predicate, maxCount, player.inventoryMenu.getCraftSlots());
-            player.containerMenu.broadcastChanges();
-            player.inventoryMenu.slotsChanged(player.getInventory());
+            // ShreddedPaper start - run on correct thread
+            CompletableFuture<Integer> future = new CompletableFuture<>();
+            ShreddedPaper.ensureSync(player, () -> {
+                int j = player.getInventory().clearOrCountMatchingItems(predicate, maxCount, player.inventoryMenu.getCraftSlots());
+                player.containerMenu.broadcastChanges();
+                player.inventoryMenu.slotsChanged(player.getInventory());
+                future.complete(j);
+            });
+            count += future.getNow(1); // ShreddedPaper - default to 1 if player is on a different thread
+            // ShreddedPaper end - run on correct thread
         }
 
         if (count == 0) {
