--- a/net/minecraft/server/commands/CloneCommands.java
+++ b/net/minecraft/server/commands/CloneCommands.java
@@ -8,9 +_,12 @@
 import com.mojang.brigadier.exceptions.Dynamic2CommandExceptionType;
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import com.mojang.logging.LogUtils;
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import java.util.Deque;
 import java.util.List;
 import java.util.function.Predicate;
+
+import io.multipaper.shreddedpaper.util.ObjectHolder;
 import net.minecraft.commands.CommandBuildContext;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
@@ -199,6 +_,7 @@
             throw ERROR_FAILED.create();
         }
 
+        ShreddedPaper.ensureSync(toDimension, destPos, () -> { // ShreddedPaper - run on correct region thread
         List<CloneCommands.CloneBlockInfo> solidList = Lists.newArrayList();
         List<CloneCommands.CloneBlockInfo> blockEntitiesList = Lists.newArrayList();
         List<CloneCommands.CloneBlockInfo> otherBlocksList = Lists.newArrayList();
@@ -298,17 +_,20 @@
                 var36.addSuppressed(var35);
             }
 
-            throw var36;
+            throw new RuntimeException(var36); // ShreddedPaper - run on correct region thread - wrap in RuntimeException for lambda
         }
 
         reporter.close();
         if (count == 0) {
-            throw ERROR_FAILED.create();
+            source.sendFailure(Component.translatable("commands.clone.failed")); // ShreddedPaper - run on correct region thread - send failure message instead of throwing
+            return; // ShreddedPaper - run on correct region thread - send failure message instead of throwing
         }
 
         int finalCount = count;
         source.sendSuccess(() -> Component.translatable("commands.clone.success", finalCount), true);
-        return count;
+        // return count; // ShreddedPaper - run on correct region thread
+        }); // ShreddedPaper - run on correct region thread
+        return 1; // ShreddedPaper - run on correct region thread
     }
 
     private record CloneBlockEntityInfo(CompoundTag tag, DataComponentMap components) {
