--- a/net/minecraft/server/commands/GiveCommand.java
+++ b/net/minecraft/server/commands/GiveCommand.java
@@ -4,6 +_,8 @@
 import com.mojang.brigadier.arguments.IntegerArgumentType;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import java.util.Collection;
+
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.commands.CommandBuildContext;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
@@ -61,8 +_,9 @@
                 int size = Math.min(maxStackSize, remaining);
                 remaining -= size;
                 ItemStack copyToDrop = prototypeItemStack.copyWithCount(size);
+                ShreddedPaper.ensureSync(player, () -> { // ShreddedPaper - run on player's thread
                 boolean added = player.getInventory().add(copyToDrop);
-                if (org.purpurmc.purpur.PurpurConfig.disableGiveCommandDrops) continue; // Purpur - add config option for toggling give command dropping
+                if (org.purpurmc.purpur.PurpurConfig.disableGiveCommandDrops) return; // Purpur - add config option for toggling give command dropping // ShreddedPaper - run on player's thread
                 if (added && copyToDrop.isEmpty()) {
                     ItemEntity drop = player.drop(prototypeItemStack.copy(), false, false, false, null); // Paper - do not fire PlayerDropItemEvent for /give command
                     if (drop != null) {
@@ -88,6 +_,7 @@
                         drop.setTarget(player.getUUID());
                     }
                 }
+                }); // ShreddedPaper - run on player's thread
             }
         }
 
