--- a/net/minecraft/util/Mth.java
+++ b/net/minecraft/util/Mth.java
@@ -317,7 +_,7 @@
         return seed >> 16;
     }
 
-    public static UUID createInsecureUUID(final RandomSource random) {
+    public static synchronized UUID createInsecureUUID(final RandomSource random) { // ShreddedPaper - Random source is not thread safe, and this is the one place it has to be thread safe
         long most = random.nextLong() & -61441L | 16384L;
         long least = random.nextLong() & 4611686018427387903L | Long.MIN_VALUE;
         return new UUID(most, least);
