--- a/src/main/java/org/bukkit/craftbukkit/util/CraftLocation.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftLocation.java
@@ -1,9 +_,11 @@
 package org.bukkit.craftbukkit.util;
 
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.SectionPos;
 import net.minecraft.core.Vec3i;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.pathfinder.Node;
 import net.minecraft.world.phys.Vec3;
@@ -72,6 +_,12 @@
         ServerLevel level = MinecraftServer.getServer().getLevel(globalPos.dimension());
         return level != null ? CraftLocation.toBukkit(globalPos.pos(), level) : null;
     }
+
+    // ShreddedPaper start
+    public static ChunkPos toChunkPosition(Location location) {
+        return new ChunkPos(SectionPos.blockToSectionCoord(location.getBlockX()), SectionPos.blockToSectionCoord(location.getBlockZ()));
+    }
+    // ShreddedPaper end
 
     public static Vec3 toVec3(Location loc) {
         return new Vec3(loc.getX(), loc.getY(), loc.getZ());
