--- a/net/minecraft/world/level/levelgen/LegacyRandomSource.java
+++ b/net/minecraft/world/level/levelgen/LegacyRandomSource.java
@@ -30,6 +_,10 @@
 
     @Override
     public void setSeed(final long seed) {
+        // ShreddedPaper start - thread safe - we don't really care about inconsistent randomness to vanilla as long as it's random
+        this.seed.set((seed ^ 25214903917L) & 281474976710655L);
+        if (true) return;
+        // ShreddedPaper end
         if (!this.seed.compareAndSet(this.seed.get(), (seed ^ 25214903917L) & 281474976710655L)) {
             throw ThreadingDetector.makeThreadingException("LegacyRandomSource", null);
         } else {
@@ -39,6 +_,14 @@
 
     @Override
     public int next(final int bits) {
+        // ShreddedPaper start - thread safe - we don't really care about inconsistent randomness to vanilla as long as it's random
+        long oldseed, nextseed;
+        do {
+            oldseed = this.seed.get();
+            nextseed = oldseed * 25214903917L + 11L & 281474976710655L;
+        } while (!this.seed.compareAndSet(oldseed, nextseed));
+        if (true) return (int) (nextseed >> 48 - bits);
+        // ShreddedPaper end
         long oldSeed = this.seed.get();
         long newSeed = oldSeed * 25214903917L + 11L & 281474976710655L;
         if (!this.seed.compareAndSet(oldSeed, newSeed)) {
