From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Caleb Graham <graham.caleb.m@gmail.com>
Date: Thu, 13 Nov 2025 01:47:08 -0600
Subject: [PATCH] Add multithreading region scheduler


diff --git a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
index 69cdd304d255d52c9b7dc9b6a33ffdb630b79abe..c9424c7b28430cc5460997603e2dc4a47b35d988 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
@@ -1,13 +1,22 @@
 package ca.spottedleaf.moonrise.common.util;
 
+import io.multipaper.shreddedpaper.threading.ShreddedPaperTickThread;
 import net.minecraft.core.BlockPos;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.Bukkit;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import io.multipaper.shreddedpaper.region.RegionPos;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
 
 import java.util.concurrent.atomic.AtomicInteger;
 
@@ -19,77 +28,104 @@ public class TickThread extends Thread {
         return "thread=" + Thread.currentThread().getName();
     }
 
+    // ShreddedPaper start
+    public static void failedTickThreadCheck(final String reason, final String extraData) {
+        if (!isTickThread()) {
+            MinecraftServer.LOGGER.error(Thread.currentThread().getName() + " tried to run a task async: " + reason);
+        } else {
+            MinecraftServer.LOGGER.error(Thread.currentThread().getName() + " tried to run a task from the wrong thread: " + reason);
+        }
+        MinecraftServer.LOGGER.error("Extra data: " + extraData);
+        MinecraftServer.LOGGER.error("Server Version: " + Bukkit.getVersion());
+        MinecraftServer.LOGGER.error("Region locks currently held (regionSize=" + RegionPos.REGION_SIZE + " chunks): " + getCurrentLocksHeld());
+        MinecraftServer.LOGGER.error("Stack trace:", new Throwable());
+        throw new IllegalStateException(reason);
+    }
+
+    private static List<String> getCurrentLocksHeld() {
+        List<String> strings = new ArrayList<>();
+
+        for (ServerLevel level : MinecraftServer.getServer().getAllLevels()) {
+            Set<RegionPos> regions = level.chunkScheduler.getRegionLocker().getLockedRegions();
+            if (!regions.isEmpty()) {
+                strings.add(level.levelStorageAccess.getLevelId() + "=" + regions.stream().map(region -> "[" + region.x + "," + region.z + "]").toList());
+            }
+        }
+
+        return strings;
+    }
+    // ShreddedPaper end
+
     /**
      * @deprecated
      */
     @Deprecated
     public static void ensureTickThread(final String reason) {
         if (!isTickThread()) {
-            LOGGER.error("Thread failed main thread check: " + reason + ", context=" + getThreadContext(), new Throwable());
-            throw new IllegalStateException(reason);
+            failedTickThreadCheck(reason, null);
         }
     }
 
     public static void ensureTickThread(final Level world, final BlockPos pos, final String reason) {
         if (!isTickThreadFor(world, pos)) {
             final String ex = "Thread failed main thread check: " +
-                               reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + pos;
+                    reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + pos;
             LOGGER.error(ex, new Throwable());
-            throw new IllegalStateException(ex);
+            failedTickThreadCheck(reason, null);
         }
     }
 
     public static void ensureTickThread(final Level world, final BlockPos pos, final int blockRadius, final String reason) {
         if (!isTickThreadFor(world, pos, blockRadius)) {
             final String ex = "Thread failed main thread check: " +
-                reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + pos + ", block_radius=" + blockRadius;
+                    reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + pos + ", block_radius=" + blockRadius;
             LOGGER.error(ex, new Throwable());
-            throw new IllegalStateException(ex);
+            failedTickThreadCheck(reason, null);
         }
     }
 
     public static void ensureTickThread(final Level world, final ChunkPos pos, final String reason) {
         if (!isTickThreadFor(world, pos)) {
             final String ex = "Thread failed main thread check: " +
-                reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", chunk_pos=" + pos;
+                    reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", chunk_pos=" + pos;
             LOGGER.error(ex, new Throwable());
-            throw new IllegalStateException(ex);
+            failedTickThreadCheck(reason, null);
         }
     }
 
     public static void ensureTickThread(final Level world, final int chunkX, final int chunkZ, final String reason) {
         if (!isTickThreadFor(world, chunkX, chunkZ)) {
             final String ex = "Thread failed main thread check: " +
-                reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", chunk_pos=" + new ChunkPos(chunkX, chunkZ);
+                    reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", chunk_pos=" + new ChunkPos(chunkX, chunkZ);
             LOGGER.error(ex, new Throwable());
-            throw new IllegalStateException(ex);
+            failedTickThreadCheck(reason, null);
         }
     }
 
     public static void ensureTickThread(final Entity entity, final String reason) {
         if (!isTickThreadFor(entity)) {
             final String ex = "Thread failed main thread check: " +
-                reason + ", context=" + getThreadContext() + ", entity=" + EntityUtil.dumpEntity(entity);
+                    reason + ", context=" + getThreadContext() + ", entity=" + EntityUtil.dumpEntity(entity);
             LOGGER.error(ex, new Throwable());
-            throw new IllegalStateException(ex);
+            failedTickThreadCheck(reason, null);
         }
     }
 
     public static void ensureTickThread(final Level world, final AABB aabb, final String reason) {
         if (!isTickThreadFor(world, aabb)) {
             final String ex = "Thread failed main thread check: " +
-                reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", aabb=" + aabb;
+                    reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", aabb=" + aabb;
             LOGGER.error(ex, new Throwable());
-            throw new IllegalStateException(ex);
+            failedTickThreadCheck(reason, null);
         }
     }
 
     public static void ensureTickThread(final Level world, final double blockX, final double blockZ, final String reason) {
         if (!isTickThreadFor(world, blockX, blockZ)) {
             final String ex = "Thread failed main thread check: " +
-                reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + new Vec3(blockX, 0.0, blockZ);
+                    reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + new Vec3(blockX, 0.0, blockZ);
             LOGGER.error(ex, new Throwable());
-            throw new IllegalStateException(ex);
+            failedTickThreadCheck(reason, null);
         }
     }
 
@@ -115,19 +151,19 @@ public class TickThread extends Thread {
     }
 
     public static TickThread getCurrentTickThread() {
-        return (TickThread)Thread.currentThread();
+        return (TickThread) Thread.currentThread();
     }
 
     public static boolean isTickThread() {
-        return Thread.currentThread() instanceof TickThread;
+        return MinecraftServer.getServer().forceTicks || isShutdownThread() || Thread.currentThread() instanceof TickThread || ShreddedPaperTickThread.isShreddedPaperTickThread(); // ShreddedPaper    }
     }
 
     public static boolean isShutdownThread() {
-        return false;
+        return Thread.currentThread() == MinecraftServer.getServer().shutdownThread; // ShreddedPaper
     }
 
     public static boolean isTickThreadFor(final Level world, final BlockPos pos) {
-        return isTickThread();
+        return isTickThreadFor(world, new ChunkPos(pos)); // ShreddedPaper
     }
 
     public static boolean isTickThreadFor(final Level world, final BlockPos pos, final int blockRadius) {
@@ -135,38 +171,72 @@ public class TickThread extends Thread {
     }
 
     public static boolean isTickThreadFor(final Level world, final ChunkPos pos) {
-        return isTickThread();
-    }
+        return ((ServerLevel) world).getServer().forceTicks || isShutdownThread() || ((ServerLevel) world).chunkScheduler.getRegionLocker().hasLock(RegionPos.forChunk(pos)); // ShreddedPaper
+        }
 
     public static boolean isTickThreadFor(final Level world, final Vec3 pos) {
-        return isTickThread();
+        return isTickThreadFor(world, pos.x, pos.z); // ShreddedPaper
     }
 
     public static boolean isTickThreadFor(final Level world, final int chunkX, final int chunkZ) {
-        return isTickThread();
+        return isTickThreadFor(world, new ChunkPos(chunkX, chunkZ)); // ShreddedPaper
     }
 
     public static boolean isTickThreadFor(final Level world, final AABB aabb) {
-        return isTickThread();
+        return isTickThreadFor(world, CoordinateUtils.getChunkCoordinate(aabb.minX), CoordinateUtils.getChunkCoordinate(aabb.minZ), CoordinateUtils.getChunkCoordinate(aabb.maxX), CoordinateUtils.getChunkCoordinate(aabb.maxZ)); // ShreddedPaper
     }
 
     public static boolean isTickThreadFor(final Level world, final double blockX, final double blockZ) {
-        return isTickThread();
+        return isTickThreadFor(world, CoordinateUtils.getChunkCoordinate(blockX), CoordinateUtils.getChunkCoordinate(blockZ)); // ShreddedPaper
     }
 
     public static boolean isTickThreadFor(final Level world, final Vec3 position, final Vec3 deltaMovement, final int buffer) {
-        return isTickThread();
+        // ShreddedPaper start
+        int fromChunkX = CoordinateUtils.getChunkCoordinate(position.x);
+        int fromChunkZ = CoordinateUtils.getChunkCoordinate(position.z);
+        int toChunkX = CoordinateUtils.getChunkCoordinate(position.x + deltaMovement.x);
+        int toChunkZ = CoordinateUtils.getChunkCoordinate(position.z + deltaMovement.z);
+
+        int minChunkX = Math.min(fromChunkX, toChunkX) - buffer;
+        int minChunkZ = Math.min(fromChunkZ, toChunkZ) - buffer;
+        int maxChunkX = Math.max(fromChunkX, toChunkX) + buffer;
+        int maxChunkZ = Math.max(fromChunkZ, toChunkZ) + buffer;
+
+        return isTickThreadFor(world, minChunkX, minChunkZ, maxChunkX, maxChunkZ);
+        // ShreddedPaper end
     }
 
     public static boolean isTickThreadFor(final Level world, final int fromChunkX, final int fromChunkZ, final int toChunkX, final int toChunkZ) {
-        return isTickThread();
+        // ShreddedPaper start
+        if (((ServerLevel) world).getServer().forceTicks || isShutdownThread()) return true;
+
+        int fromRegionX = fromChunkX >> RegionPos.REGION_SHIFT;
+        int fromRegionZ = fromChunkZ >> RegionPos.REGION_SHIFT;
+        int toRegionX = toChunkX >> RegionPos.REGION_SHIFT;
+        int toRegionZ = toChunkZ >> RegionPos.REGION_SHIFT;
+
+        int minRegionX = Math.min(fromRegionX, toRegionX);
+        int maxRegionX = Math.max(fromRegionX, toRegionX);
+        int minRegionZ = Math.min(fromRegionZ, toRegionZ);
+        int maxRegionZ = Math.max(fromRegionZ, toRegionZ);
+
+        for (int x = minRegionX; x <= maxRegionX; x++) {
+            for (int z = minRegionZ; z <= maxRegionZ; z++) {
+                if (!((ServerLevel) world).chunkScheduler.getRegionLocker().hasLock(new RegionPos(x, z))) {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+        // ShreddedPaper end
     }
 
     public static boolean isTickThreadFor(final Level world, final int chunkX, final int chunkZ, final int radius) {
-        return isTickThread();
+        return isTickThreadFor(world, chunkX - radius, chunkZ - radius, chunkX + radius, chunkZ + radius); // ShreddedPaper
     }
 
     public static boolean isTickThreadFor(final Entity entity) {
-        return isTickThread();
+        return isTickThreadFor(entity.level(), entity.chunkPosition()); // ShreddedPaper
     }
 }
diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
index 7ce9ebba8ce304d1f3f21d4f15ee5f3560d7700b..a9a74db4db3667d7186d79a7aac5dd95a821ff7b 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
@@ -1,9 +1,11 @@
 package io.papermc.paper.plugin.manager;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import co.aikar.timings.TimedEventExecutor;
 import com.destroystokyo.paper.event.server.ServerExceptionEvent;
 import com.destroystokyo.paper.exception.ServerEventException;
 import com.google.common.collect.Sets;
+import io.multipaper.shreddedpaper.threading.ShreddedPaperTickThread;
 import org.bukkit.Server;
 import org.bukkit.Warning;
 import org.bukkit.event.Event;
@@ -36,7 +38,7 @@ class PaperEventManager {
 
     // SimplePluginManager
     public void callEvent(@NotNull Event event) {
-        if (event.isAsynchronous() && this.server.isPrimaryThread()) {
+        if (event.isAsynchronous() && (Thread.currentThread() instanceof TickThread || ShreddedPaperTickThread.isShreddedPaperTickThread())) {
             throw new IllegalStateException(event.getEventName() + " may only be triggered asynchronously.");
         } else if (!event.isAsynchronous() && !this.server.isPrimaryThread() && !this.server.isStopping()) {
             throw new IllegalStateException(event.getEventName() + " may only be triggered synchronously.");
