From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: caleb <graham.caleb.m@gmail.com>
Date: Mon, 1 Dec 2025 01:14:30 -0600
Subject: [PATCH] Ensure-worlds-are-loaded-on-the-global-scheduler


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 78eb132db5ed6e4c09759cb86a033d3f9e7310dc..cbd0d3af7797cfe7f44b32a54824593cc79cb3c3 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1252,6 +1252,8 @@ public final class CraftServer implements Server {
         Preconditions.checkState(this.console.getAllLevels().iterator().hasNext(), "Cannot create additional worlds on STARTUP");
         //Preconditions.checkState(!this.console.isIteratingOverLevels, "Cannot create a world while worlds are being ticked"); // Paper - Cat - Temp disable. We'll see how this goes.
         Preconditions.checkArgument(creator != null, "WorldCreator cannot be null");
+        if (!TickThread.isTickThread()) throw new IllegalStateException("Cannot load worlds async!!"); // ShreddedPaper
+        if (ShreddedPaperTickThread.isShreddedPaperTickThread()) throw new IllegalStateException("Must load worlds from the global scheduler!"); // ShreddedPaper
 
         String name = creator.name();
         ChunkGenerator chunkGenerator = creator.generator();
