From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: caleb <graham.caleb.m@gmail.com>
Date: Mon, 1 Dec 2025 01:42:07 -0600
Subject: [PATCH] Bukkit-API-thread-checks


diff --git a/src/main/java/ca/spottedleaf/moonrise/common/util/ChunkSystemHooks.java b/src/main/java/ca/spottedleaf/moonrise/common/util/ChunkSystemHooks.java
index 15a3c2436b04f0e1b2a538682183ff3ce9c0573e..2483ff8dbc207d11ffe28eea106bd3487e90715a 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/util/ChunkSystemHooks.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/ChunkSystemHooks.java
@@ -20,8 +20,8 @@ public interface ChunkSystemHooks {
     public void scheduleChunkTask(final ServerLevel level, final int chunkX, final int chunkZ, final Runnable run, final Priority priority);
 
     public void scheduleChunkLoad(final ServerLevel level, final int chunkX, final int chunkZ, final boolean gen,
-                                  final ChunkStatus toStatus, final boolean addTicket, final Priority priority,
-                                  final Consumer<ChunkAccess> onComplete);
+                                         final ChunkStatus toStatus, final boolean addTicket, final Priority priority,
+                                         final Consumer<ChunkAccess> onComplete);
 
     public void scheduleChunkLoad(final ServerLevel level, final int chunkX, final int chunkZ, final ChunkStatus toStatus,
                                   final boolean addTicket, final Priority priority, final Consumer<ChunkAccess> onComplete);
diff --git a/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitCommandNode.java b/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitCommandNode.java
index a127e982e6ee1bf035ce343a6751c8894d4d93c9..427698ee890c09fbf70bd8ed4700870759e560ad 100644
--- a/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitCommandNode.java
+++ b/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitCommandNode.java
@@ -137,6 +137,7 @@ public class BukkitCommandNode extends LiteralCommandNode<CommandSourceStack> {
 
             if (sender instanceof final Player player) {
                 TabCompleteEvent tabEvent = new org.bukkit.event.server.TabCompleteEvent(player, builder.getInput(), results.value() != null ? results.value() : new ArrayList<>(), true, pos); // Paper - AsyncTabCompleteEvent // ShreddedPaper
+                if (!tabEvent.callEvent()) {
                     results.value(null); // ShreddedPaper
                 } else {
                     results.value(tabEvent.getCompletions()); // ShreddedPaper
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index 400e632208d133a3f49fc7f14bceb48a1026769b..6365f230512a27cc47cffde913988454d0e0dea9 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -1,5 +1,6 @@
 package org.bukkit.craftbukkit;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import com.google.common.base.Preconditions;
 import com.google.common.base.Predicates;
 import com.mojang.serialization.Codec;
@@ -9,6 +10,8 @@ import java.util.Arrays;
 import java.util.Collection;
 import java.util.List;
 import java.util.function.Predicate;
+
+import io.multipaper.shreddedpaper.threading.ShreddedPaperTickThread;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.Registry;
@@ -77,6 +80,7 @@ public class CraftChunk implements Chunk {
     }
 
     public ChunkAccess getHandle(ChunkStatus chunkStatus) {
+        if (ShreddedPaperTickThread.isShreddedPaperTickThread() && !TickThread.isTickThreadFor(this.level, this.x, this.z)) TickThread.failedTickThreadCheck("Cannot get chunk from a region that is not ours!", "world=" + this.level.levelStorageAccess.getLevelId() + ", chunkpos=[" + this.x + "," + this.z + "]"); // ShreddedPaper - regions
         // Paper start - chunk system
         net.minecraft.world.level.chunk.LevelChunk full = this.level.getChunkIfLoaded(this.x, this.z);
         if (full != null) {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 8fd6c341671eef2edfa95824f9d91aa7d235c016..034da58f5b85b07ff70de16892ee5154e0c69850 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -3,11 +3,13 @@ package org.bukkit.craftbukkit;
 import ca.spottedleaf.moonrise.common.list.ReferenceList;
 import ca.spottedleaf.moonrise.common.util.CoordinateUtils;
 import ca.spottedleaf.moonrise.common.util.TickThread;
+import ca.spottedleaf.moonrise.paper.util.BaseChunkSystemHooks;
 import com.google.common.base.Preconditions;
 import com.google.common.base.Predicates;
 import com.mojang.datafixers.util.Pair;
 import io.multipaper.shreddedpaper.ShreddedPaper;
 import io.multipaper.shreddedpaper.region.RegionPos;
+import io.multipaper.shreddedpaper.threading.ShreddedPaperTickThread;
 import io.multipaper.shreddedpaper.threading.SynchronousPluginExecution;
 import io.papermc.paper.FeatureHooks;
 import io.papermc.paper.raytracing.RayTraceTarget;
@@ -382,6 +384,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
     @Override
     public Chunk getChunkAt(int x, int z) {
+        if (ShreddedPaperTickThread.isShreddedPaperTickThread() && !TickThread.isTickThreadFor(this.world, x, z)) TickThread.failedTickThreadCheck("Cannot get chunk from a region that is not ours!", "world=" + this.world.getLevel() + ", chunkpos=[" + x + "," + z + "]"); // ShreddedPaper - regions
         warnUnsafeChunk("getting a faraway chunk", x, z); // Paper
         net.minecraft.world.level.chunk.LevelChunk chunk = (net.minecraft.world.level.chunk.LevelChunk) this.world.getChunk(x, z, ChunkStatus.FULL, true);
         return new CraftChunk(chunk);
@@ -425,7 +428,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         }
         final java.util.concurrent.CompletableFuture<ChunkAccess> future = new java.util.concurrent.CompletableFuture<>();
         ca.spottedleaf.moonrise.common.PlatformHooks.get().scheduleChunkLoad(
-            this.world, x, z, false, ChunkStatus.EMPTY, true, ca.spottedleaf.concurrentutil.util.Priority.NORMAL, future::complete
+            this.world, x, z,  ChunkStatus.EMPTY, true, ca.spottedleaf.concurrentutil.util.Priority.NORMAL, future::complete
         );
         world.getChunkSource().mainThreadProcessor.managedBlock(future::isDone);
         return future.thenApply(c -> {
@@ -437,6 +440,8 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         // Paper end - Fix this method
     }
 
+
+
     @Override
     public Chunk[] getLoadedChunks() {
         ServerChunkCache serverChunkCache = this.getHandle().chunkSource;
@@ -2495,17 +2500,23 @@ public class CraftWorld extends CraftRegionAccessor implements World {
     @Override
     public void getChunkAtAsync(int x, int z, boolean gen, boolean urgent, @NotNull Consumer<? super Chunk> cb) {
         warnUnsafeChunk("getting a faraway chunk async", x, z); // Paper
-            java.util.concurrent.CompletableFuture<Chunk> ret = new java.util.concurrent.CompletableFuture<>();  // ShreddedPaper - synchronous plugin execution
-            Plugin plugin = SynchronousPluginExecution.getCurrentPlugin(); // ShreddedPaper - synchronous plugin execution
-            ShreddedPaper.ensureSync(this.getHandle(), c.getPos(), () -> { // ShreddedPaper - ensure on correct thread
-                this.getHandle(), x, z, gen, ChunkStatus.FULL, true,
-                        urgent ? ca.spottedleaf.concurrentutil.util.Priority.HIGHER : ca.spottedleaf.concurrentutil.util.Priority.NORMAL,
-                        (ChunkAccess chunk) -> {
-                            SynchronousPluginExecution.executeNoException(plugin, () -> ret.complete(chunk == null ? null : new CraftChunk(chunk))); // ShreddedPaper - synchronous plugin execution
-                        }
-            }
-        );
+        Plugin plugin = SynchronousPluginExecution.getCurrentPlugin(); // ShreddedPaper - synchronous plugin execution
 
+        ca.spottedleaf.moonrise.common.PlatformHooks.get().scheduleChunkLoad(
+                this.getHandle().getLevel(),
+                x,
+                z,
+                gen,
+                ChunkStatus.FULL,
+                true,
+                urgent ? ca.spottedleaf.concurrentutil.util.Priority.HIGHER : ca.spottedleaf.concurrentutil.util.Priority.NORMAL,
+                (ChunkAccess c) -> {
+                    ShreddedPaper.ensureSync(this.getHandle(), c.getPos(), () -> { // ShreddedPaper - ensure on correct thread
+                        net.minecraft.world.level.chunk.LevelChunk chunk = (net.minecraft.world.level.chunk.LevelChunk) c;
+                        SynchronousPluginExecution.executeNoException(plugin, () -> cb.accept(chunk == null ? null : new CraftChunk(chunk))); // ShreddedPaper - synchronous plugin execution
+                    });
+                }
+        );
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index 054217727626193a8f402762fb5d7d8cf417b717..5489f90184d3ea844f687afd5c7bdc7e45cbc8dc 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -1,5 +1,6 @@
 package org.bukkit.craftbukkit.block;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import com.google.common.base.Preconditions;
 import java.util.ArrayList;
 import java.util.Collection;
@@ -75,6 +76,7 @@ public class CraftBlock implements Block {
     }
 
     public net.minecraft.world.level.block.state.BlockState getNMS() {
+        if (this.world instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, this.position, "Cannot get block asynchronously!");
         return this.world.getBlockState(this.position);
     }
 
@@ -160,6 +162,7 @@ public class CraftBlock implements Block {
 
     @Override
     public byte getData() {
+        if (this.world instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, this.position, "Cannot get block asynchronously!");
         net.minecraft.world.level.block.state.BlockState state = this.world.getBlockState(this.position);
         return CraftMagicNumbers.toLegacyData(state);
     }
@@ -192,6 +195,7 @@ public class CraftBlock implements Block {
     }
 
     boolean setBlockState(final net.minecraft.world.level.block.state.BlockState state, final boolean applyPhysics) {
+        if (this.world instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, this.position, "Cannot set block asynchronously!");
         return CraftBlock.setBlockState(this.world, this.position, this.getNMS(), state, applyPhysics);
     }
 
@@ -227,6 +231,7 @@ public class CraftBlock implements Block {
 
     @Override
     public Material getType() {
+        if (this.world instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, this.position, "Cannot get block asynchronously!");
         return this.world.getBlockState(this.position).getBukkitMaterial(); // Paper - optimise getType calls
     }
 
