From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: caleb <graham.caleb.m@gmail.com>
Date: Sun, 26 Oct 2025 14:41:40 -0500
Subject: [PATCH] Syncronization and locks


diff --git a/ca/spottedleaf/moonrise/common/list/ReferenceList.java b/ca/spottedleaf/moonrise/common/list/ReferenceList.java
index 2e876b918672e8ef3b5197b7e6b1597247fdeaa1..940920ff4425625ca319e55e679fe93c822a1fad 100644
--- a/ca/spottedleaf/moonrise/common/list/ReferenceList.java
+++ b/ca/spottedleaf/moonrise/common/list/ReferenceList.java
@@ -4,6 +4,7 @@ import it.unimi.dsi.fastutil.objects.Reference2IntOpenHashMap;
 import java.util.Arrays;
 import java.util.Iterator;
 import java.util.NoSuchElementException;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 public final class ReferenceList<E> implements Iterable<E> {
 
@@ -12,6 +13,7 @@ public final class ReferenceList<E> implements Iterable<E> {
     private final Reference2IntOpenHashMap<E> referenceToIndex;
     private E[] references;
     private int count;
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
 
     public ReferenceList() {
         this((E[])EMPTY_LIST);
@@ -30,80 +32,130 @@ public final class ReferenceList<E> implements Iterable<E> {
     }
 
     public ReferenceList<E> copy() {
-        return new ReferenceList<>(this.references.clone(), this.count, this.referenceToIndex.clone());
+        lock.readLock().lock();
+        try {
+            return new ReferenceList<>(this.references.clone(), this.count, this.referenceToIndex.clone());
+        } finally {
+            lock.readLock().unlock();
+        }
     }
 
     public int size() {
-        return this.count;
+        lock.readLock().lock();
+        try {
+            return this.count;
+        } finally {
+        lock.readLock().unlock();
+        }
     }
 
     public boolean contains(final E obj) {
-        return this.referenceToIndex.containsKey(obj);
+        lock.readLock().lock();
+        try {
+            return this.referenceToIndex.containsKey(obj);
+        } finally {
+            lock.readLock().unlock();
+        }
     }
 
     public boolean remove(final E obj) {
-        final int index = this.referenceToIndex.removeInt(obj);
-        if (index == Integer.MIN_VALUE) {
-            return false;
-        }
+        lock.writeLock().lock();
+        try {
+            final int index = this.referenceToIndex.removeInt(obj);
+            if (index == Integer.MIN_VALUE) {
+                return false;
+            }
 
-        // move the object at the end to this index
-        final int endIndex = --this.count;
-        final E end = (E)this.references[endIndex];
-        if (index != endIndex) {
-            // not empty after this call
-            this.referenceToIndex.put(end, index); // update index
-        }
-        this.references[index] = end;
-        this.references[endIndex] = null;
+            // move the object at the end to this index
+            final int endIndex = --this.count;
+            final E end = (E) this.references[endIndex];
+            if (index != endIndex) {
+                // not empty after this call
+                this.referenceToIndex.put(end, index); // update index
+            }
+            this.references[index] = end;
+            this.references[endIndex] = null;
 
-        return true;
+            return true;
+        } finally {
+            lock.writeLock().unlock();
+        }
     }
 
     public boolean add(final E obj) {
-        final int count = this.count;
-        final int currIndex = this.referenceToIndex.putIfAbsent(obj, count);
+        lock.writeLock().lock();
+        try {
+            final int count = this.count;
+            final int currIndex = this.referenceToIndex.putIfAbsent(obj, count);
 
-        if (currIndex != Integer.MIN_VALUE) {
-            return false; // already in this list
-        }
+            if (currIndex != Integer.MIN_VALUE) {
+                return false; // already in this list
+            }
 
-        E[] list = this.references;
+            E[] list = this.references;
 
-        if (list.length == count) {
-            // resize required
-            list = this.references = Arrays.copyOf(list, (int)Math.max(4L, count * 2L)); // overflow results in negative
-        }
+            if (list.length == count) {
+                // resize required
+                list = this.references = Arrays.copyOf(list, (int) Math.max(4L, count * 2L)); // overflow results in negative
+            }
 
-        list[count] = obj;
-        this.count = count + 1;
+            list[count] = obj;
+            this.count = count + 1;
 
-        return true;
+            return true;
+        } finally {
+            lock.writeLock().unlock();
+        }
     }
 
     public E getChecked(final int index) {
-        if (index < 0 || index >= this.count) {
-            throw new IndexOutOfBoundsException("Index: " + index + " is out of bounds, size: " + this.count);
+        lock.readLock().lock();
+        try {
+            if (index < 0 || index >= this.count) {
+                throw new IndexOutOfBoundsException("Index: " + index + " is out of bounds, size: " + this.count);
+            }
+            return this.references[index];
+        } finally {
+            lock.readLock().unlock();
         }
-        return this.references[index];
     }
 
     public E getUnchecked(final int index) {
-        return this.references[index];
+        lock.readLock().lock();
+        try {
+            return this.references[index];
+        } finally {
+            lock.readLock().unlock();
+        }
     }
 
     public Object[] getRawData() {
-        return this.references;
+        lock.readLock().lock();
+        try {
+            return this.references;
+        } finally {
+            lock.readLock().unlock();
+        }
     }
 
     public E[] getRawDataUnchecked() {
-        return this.references;
+        lock.readLock().lock();
+        try {
+            return this.references;
+        } finally {
+            lock.readLock().unlock();
+        }
     }
 
     public void clear() {
-        this.referenceToIndex.clear();
-        Arrays.fill(this.references, 0, this.count, null);
-        this.count = 0;
+        lock.writeLock().lock();
+        try {
+            this.referenceToIndex.clear();
+            Arrays.fill(this.references, 0, this.count, null);
+            this.count = 0;
+        } finally {
+            lock.writeLock().unlock();
+        }
     }
 
     @Override
@@ -112,14 +164,22 @@ public final class ReferenceList<E> implements Iterable<E> {
             private E lastRet;
             private int current;
 
+            {
+                lock.readLock().lock();
+            }
             @Override
             public boolean hasNext() {
-                return this.current < ReferenceList.this.count;
+                boolean hasNext = this.current < ReferenceList.this.count;
+                if (!hasNext){
+                    lock.readLock().unlock();
+                }
+                return hasNext;
             }
 
             @Override
             public E next() {
                 if (this.current >= ReferenceList.this.count) {
+                    lock.readLock().unlock();
                     throw new NoSuchElementException();
                 }
                 return this.lastRet = ReferenceList.this.references[this.current++];
@@ -134,8 +194,16 @@ public final class ReferenceList<E> implements Iterable<E> {
                 }
                 this.lastRet = null;
 
-                ReferenceList.this.remove(lastRet);
-                --this.current;
+                lock.readLock().unlock();
+                lock.writeLock().lock();
+
+                try {
+                    ReferenceList.this.remove(lastRet);
+                    --this.current;
+                } finally {
+                    lock.writeLock().unlock();
+                    lock.readLock().lock();
+                }
             }
         };
     }
diff --git a/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkChangesBroadcaster.java b/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkChangesBroadcaster.java
index 3bcef937b556d152ce24d50ec1924a08c5f4e5ef..9128629bfa167215381cd7073b72d9e2834d7e12 100644
--- a/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkChangesBroadcaster.java
+++ b/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkChangesBroadcaster.java
@@ -15,7 +15,7 @@ public class ShreddedPaperChunkChangesBroadcaster {
     private static final Logger LOGGER = LogUtils.getLogger();
 
     private static final Set<LevelChunk> globalToBroadcast = ConcurrentHashMap.newKeySet();
-    private static final ThreadLocal<Set<LevelChunk>> thisThreadToBroadcast = ThreadLocal.withInitial(null);
+    private static final ThreadLocal<Set<LevelChunk>> thisThreadToBroadcast = ThreadLocal.withInitial(() -> null);
 
     public static void setNeedsBroadcast(LevelChunk chunk) {
         if (thisThreadToBroadcast.get() != null && TickThread.isTickThreadFor(chunk.getLevel(), chunk.getPos())) {
diff --git a/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java b/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
index 51ede3e128e572d40b9e412eccecbb312fef757e..57c8aeb70bf394f17faf23350d913d27ec513130 100644
--- a/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
+++ b/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
@@ -80,7 +80,7 @@ public class ShreddedPaperChunkTicker {
 
             while (region.getInternalTaskQueue().executeTask()) ;
 
-            level.moonrise$getChunkTaskScheduler().chunkHolderManager.processUnloads(region);
+            //level.moonrise$getChunkTaskScheduler().chunkHolderManager.processUnloads(region);
 
             region.forEachTickingEntity(entity -> {
                 CraftEntity bukkitEntity = entity.getBukkitEntityRaw();
@@ -102,7 +102,13 @@ public class ShreddedPaperChunkTicker {
 
             level.handlingTickThreadLocal.set(false);
 
-            region.forEachTickingEntity(ShreddedPaperEntityTicker::tickEntity);
+            region.forEachTickingEntity(entity -> {
+                synchronized (entity) {
+                    if (!entity.isRemoved()) {
+                        ShreddedPaperEntityTicker.tickEntity(entity);
+                    }
+                }
+            });
 
             region.forEachTrackedEntity(ShreddedPaperEntityTicker::processTrackQueue);
 
diff --git a/io/multipaper/shreddedpaper/threading/ShreddedPaperEntityTicker.java b/io/multipaper/shreddedpaper/threading/ShreddedPaperEntityTicker.java
index ab51946555b14c60a633bb29ca389837f9d8764c..e851b4fa86949c13f477e7da3a5b559497586cc5 100644
--- a/io/multipaper/shreddedpaper/threading/ShreddedPaperEntityTicker.java
+++ b/io/multipaper/shreddedpaper/threading/ShreddedPaperEntityTicker.java
@@ -4,6 +4,7 @@ import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ChunkMap;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.Mob;
 
 import java.util.Objects;
 
@@ -14,38 +15,43 @@ public class ShreddedPaperEntityTicker {
 
         entity.activatedPriorityReset = false; // Pufferfish - DAB
         if (!entity.isRemoved()) {
-            if (false && level.shouldDiscardEntity(entity)) { // CraftBukkit - We prevent spawning in general, so this butchering is not needed
-                entity.discard();
-            } else if (!level.tickRateManager().isEntityFrozen(entity)) {
-                //gameprofilerfiller.push("checkDespawn"); // Purpur
-                entity.checkDespawn();
-                //gameprofilerfiller.pop(); // Purpur
-                if (true || level.chunkSource.chunkMap.getDistanceManager().inEntityTickingRange(entity.chunkPosition().toLong())) { // Paper - rewrite chunk system
-                    Entity entity1 = entity.getVehicle();
-
-                    if (entity1 != null) {
-                        if (!entity1.isRemoved() && entity1.hasPassenger(entity)) {
-                            return;
-                        }
-
-                        entity.stopRiding();
+            synchronized (entity) {
+                if (false && level.shouldDiscardEntity(entity)) { // CraftBukkit - We prevent spawning in general, so this butchering is not needed
+                    entity.discard();
+                } else if (!level.tickRateManager().isEntityFrozen(entity)) {
+                    //gameprofilerfiller.push("checkDespawn"); // Purpur
+                    if (!entity.isRemoved() && entity instanceof Mob mob) {
+                        mob.checkDespawn();
                     }
+                    entity.checkDespawn();
+                    //gameprofilerfiller.pop(); // Purpur
+                    if (true || level.chunkSource.chunkMap.getDistanceManager().inEntityTickingRange(entity.chunkPosition().toLong())) { // Paper - rewrite chunk system
+                        Entity entity1 = entity.getVehicle();
+
+                        if (entity1 != null) {
+                            if (!entity1.isRemoved() && entity1.hasPassenger(entity)) {
+                                return;
+                            }
 
-                    //gameprofilerfiller.push("tick"); // Purpur
-                    // Pufferfish start - copied from this.guardEntityTick
-                    try {
-                        level.tickNonPassenger(entity); // Pufferfish - changed
-                    } catch (Throwable throwable) {
-                        if (throwable instanceof ThreadDeath) throw throwable; // Paper
-                        // Paper start - Prevent tile entity and entity crashes
-                        final String msg = String.format("Entity threw exception at %s:%s,%s,%s", entity.level().getWorld().getName(), entity.getX(), entity.getY(), entity.getZ());
-                        MinecraftServer.LOGGER.error(msg, throwable);
-                        level.getCraftServer().getPluginManager().callEvent(new com.destroystokyo.paper.event.server.ServerExceptionEvent(new com.destroystokyo.paper.exception.ServerInternalException(msg, throwable)));
-                        entity.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD);
-                        // Paper end
+                            entity.stopRiding();
+                        }
+
+                        //gameprofilerfiller.push("tick"); // Purpur
+                        // Pufferfish start - copied from this.guardEntityTick
+                        try {
+                            level.tickNonPassenger(entity); // Pufferfish - changed
+                        } catch (Throwable throwable) {
+                            if (throwable instanceof ThreadDeath) throw throwable; // Paper
+                            // Paper start - Prevent tile entity and entity crashes
+                            final String msg = String.format("Entity threw exception at %s:%s,%s,%s", entity.level().getWorld().getName(), entity.getX(), entity.getY(), entity.getZ());
+                            MinecraftServer.LOGGER.error(msg, throwable);
+                            level.getCraftServer().getPluginManager().callEvent(new com.destroystokyo.paper.event.server.ServerExceptionEvent(new com.destroystokyo.paper.exception.ServerInternalException(msg, throwable)));
+                            entity.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD);
+                            // Paper end
+                        }
+                        // Pufferfish end
+                        //gameprofilerfiller.pop(); // Purpur
                     }
-                    // Pufferfish end
-                    //gameprofilerfiller.pop(); // Purpur
                 }
             }
         }
