From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: caleb <graham.caleb.m@gmail.com>
Date: Thu, 30 Oct 2025 04:21:13 -0500
Subject: [PATCH] Fix Chunk Sending


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index 4de26e3e46f131aea9dc893d67f6e108c1ad303d..1cc56c9d8734ac073a0086df21af9a9d8926270e 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -14,6 +14,7 @@ import ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkHolderManage
 import ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler;
 import ca.spottedleaf.moonrise.patches.chunk_system.util.ParallelSearchRadiusIteration;
 import com.google.gson.JsonObject;
+import com.mojang.logging.LogUtils;
 import io.multipaper.shreddedpaper.ShreddedPaper;
 import io.multipaper.shreddedpaper.region.RegionPos;
 import io.multipaper.shreddedpaper.threading.ShreddedPaperChunkTicker;
@@ -38,6 +39,7 @@ import net.minecraft.world.level.chunk.ChunkAccess;
 import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.chunk.status.ChunkStatus;
 import net.minecraft.world.level.levelgen.BelowZeroRetrogen;
+
 import java.lang.invoke.VarHandle;
 import java.util.ArrayDeque;
 import java.util.concurrent.TimeUnit;
@@ -56,6 +58,8 @@ public final class RegionizedPlayerChunkLoader {
     public static final int LOADED_TICKET_LEVEL = ChunkTaskScheduler.getTicketLevel(ChunkStatus.EMPTY);
     public static final int TICK_TICKET_LEVEL = ChunkHolderManager.ENTITY_TICKING_TICKET_LEVEL;
 
+
+
     public static final class ViewDistanceHolder {
 
         private volatile ViewDistances viewDistances;
@@ -656,7 +660,7 @@ public final class RegionizedPlayerChunkLoader {
                 final long pendingLoadChunk = this.loadingQueue.firstLong();
                 final int pendingChunkX = CoordinateUtils.getChunkX(pendingLoadChunk);
                 final int pendingChunkZ = CoordinateUtils.getChunkZ(pendingLoadChunk);
-                final ChunkAccess pending = ((ChunkSystemLevel)this.world).moonrise$getAnyChunkIfLoaded(pendingChunkX, pendingChunkZ);
+                final ChunkAccess pending = ((ChunkSystemLevel) this.world).moonrise$getAnyChunkIfLoaded(pendingChunkX, pendingChunkZ);
                 if (pending == null) {
                     // nothing to do here
                     break;
@@ -677,7 +681,7 @@ public final class RegionizedPlayerChunkLoader {
 
             // try to push more chunk loads
             final long maxLoads = Math.max(0L, Math.min(MAX_RATE, Math.min(this.loadQueue.size(), this.getMaxChunkLoads())));
-            final int maxLoadsThisTick = (int)this.chunkLoadTicketLimiter.takeAllocation(time, loadRate, maxLoads);
+            final int maxLoadsThisTick = (int) this.chunkLoadTicketLimiter.takeAllocation(time, loadRate, maxLoads);
             if (maxLoadsThisTick > 0) {
                 final LongArrayList chunks = new LongArrayList(maxLoadsThisTick);
                 for (int i = 0; i < maxLoadsThisTick; ++i) {
@@ -687,10 +691,10 @@ public final class RegionizedPlayerChunkLoader {
                         throw new IllegalStateException("Previous state should be " + CHUNK_TICKET_STAGE_NONE + ", not " + prev);
                     }
                     this.pushDelayedTicketOp(
-                        ChunkHolderManager.TicketOperation.addOp(
-                            chunk,
-                            PLAYER_TICKET, LOADED_TICKET_LEVEL, this.idBoxed
-                        )
+                            ChunkHolderManager.TicketOperation.addOp(
+                                    chunk,
+                                    PLAYER_TICKET, LOADED_TICKET_LEVEL, this.idBoxed
+                            )
                     );
                     chunks.add(chunk);
                     this.loadingQueue.enqueue(chunk);
@@ -702,7 +706,7 @@ public final class RegionizedPlayerChunkLoader {
                 // load - only generate ticket levels start anything, but they start generation...
                 // propagate levels
                 // Note: this CAN call plugin logic, so it is VITAL that our bookkeeping logic is completely done by the time this is invoked
-                ((ChunkSystemServerLevel)this.world).moonrise$getChunkTaskScheduler().chunkHolderManager.processTicketUpdates();
+                ((ChunkSystemServerLevel) this.world).moonrise$getChunkTaskScheduler().chunkHolderManager.processTicketUpdates();
 
                 if (this.removed) {
                     // process ticket updates may invoke plugin logic, which may remove this player
@@ -713,8 +717,8 @@ public final class RegionizedPlayerChunkLoader {
                     final long queuedLoadChunk = chunks.getLong(i);
                     final int queuedChunkX = CoordinateUtils.getChunkX(queuedLoadChunk);
                     final int queuedChunkZ = CoordinateUtils.getChunkZ(queuedLoadChunk);
-                    ((ChunkSystemServerLevel)this.world).moonrise$getChunkTaskScheduler().scheduleChunkLoad(
-                        queuedChunkX, queuedChunkZ, ChunkStatus.EMPTY, false, PrioritisedExecutor.Priority.NORMAL, null
+                    ((ChunkSystemServerLevel) this.world).moonrise$getChunkTaskScheduler().scheduleChunkLoad(
+                            queuedChunkX, queuedChunkZ, ChunkStatus.EMPTY, false, PrioritisedExecutor.Priority.NORMAL, null
                     );
                     if (this.removed) {
                         return;
@@ -727,7 +731,7 @@ public final class RegionizedPlayerChunkLoader {
                 final long pendingGenChunk = this.generatingQueue.firstLong();
                 final int pendingChunkX = CoordinateUtils.getChunkX(pendingGenChunk);
                 final int pendingChunkZ = CoordinateUtils.getChunkZ(pendingGenChunk);
-                final LevelChunk pending = ((ChunkSystemLevel)this.world).moonrise$getFullChunkIfLoaded(pendingChunkX, pendingChunkZ);
+                final LevelChunk pending = ((ChunkSystemLevel) this.world).moonrise$getFullChunkIfLoaded(pendingChunkX, pendingChunkZ);
                 if (pending == null) {
                     // nothing to do here
                     break;
@@ -760,7 +764,7 @@ public final class RegionizedPlayerChunkLoader {
                 final long chunkKey = this.genQueue.firstLong();
                 final int chunkX = CoordinateUtils.getChunkX(chunkKey);
                 final int chunkZ = CoordinateUtils.getChunkZ(chunkKey);
-                final ChunkAccess chunk = ((ChunkSystemLevel)this.world).moonrise$getAnyChunkIfLoaded(chunkX, chunkZ);
+                final ChunkAccess chunk = ((ChunkSystemLevel) this.world).moonrise$getAnyChunkIfLoaded(chunkX, chunkZ);
                 if (chunk.getPersistedStatus() != ChunkStatus.FULL) {
                     // only rate limit actual generations
                     if ((ratedGensThisTick + 1L) > maxGensThisTick) {
@@ -776,11 +780,11 @@ public final class RegionizedPlayerChunkLoader {
                     throw new IllegalStateException("Previous state should be " + CHUNK_TICKET_STAGE_LOADED + ", not " + prev);
                 }
                 this.pushDelayedTicketOp(
-                    ChunkHolderManager.TicketOperation.addAndRemove(
-                            chunkKey,
-                            PLAYER_TICKET, GENERATED_TICKET_LEVEL, this.idBoxed,
-                            PLAYER_TICKET, LOADED_TICKET_LEVEL, this.idBoxed
-                    )
+                        ChunkHolderManager.TicketOperation.addAndRemove(
+                                chunkKey,
+                                PLAYER_TICKET, GENERATED_TICKET_LEVEL, this.idBoxed,
+                                PLAYER_TICKET, LOADED_TICKET_LEVEL, this.idBoxed
+                        )
                 );
                 this.generatingQueue.enqueue(chunkKey);
             }
@@ -794,18 +798,18 @@ public final class RegionizedPlayerChunkLoader {
                 final int pendingChunkZ = CoordinateUtils.getChunkZ(pendingTicking);
 
                 if (!this.areNeighboursGenerated(pendingChunkX, pendingChunkZ,
-                    ChunkHolderManager.FULL_LOADED_TICKET_LEVEL - ChunkHolderManager.ENTITY_TICKING_TICKET_LEVEL)) {
+                        ChunkHolderManager.FULL_LOADED_TICKET_LEVEL - ChunkHolderManager.ENTITY_TICKING_TICKET_LEVEL)) {
                     break;
                 }
 
                 // only gets here if all neighbours were marked as generated or ticking themselves
                 this.tickingQueue.dequeueLong();
                 this.pushDelayedTicketOp(
-                    ChunkHolderManager.TicketOperation.addAndRemove(
-                            pendingTicking,
-                            PLAYER_TICKET, TICK_TICKET_LEVEL, this.idBoxed,
-                            PLAYER_TICKET, GENERATED_TICKET_LEVEL, this.idBoxed
-                    )
+                        ChunkHolderManager.TicketOperation.addAndRemove(
+                                pendingTicking,
+                                PLAYER_TICKET, TICK_TICKET_LEVEL, this.idBoxed,
+                                PLAYER_TICKET, GENERATED_TICKET_LEVEL, this.idBoxed
+                        )
                 );
                 // note: there is no queue to add after ticking
                 final byte prev = this.chunkTicketStage.put(pendingTicking, CHUNK_TICKET_STAGE_TICK);
@@ -816,63 +820,78 @@ public final class RegionizedPlayerChunkLoader {
 
             // try to pull sending chunks
             final long maxSends = Math.max(0L, Math.min(MAX_RATE, Integer.MAX_VALUE)); // note: no logic to track concurrent sends
-            final int maxSendsThisTick = Math.min((int)this.chunkSendLimiter.takeAllocation(time, sendRate, maxSends), this.sendQueue.size());
+            final int maxSendsThisTick = Math.min((int) this.chunkSendLimiter.takeAllocation(time, sendRate, maxSends), this.sendQueue.size());
             // we do not return sends that we took from the allocation back because we want to limit the max send rate, not target it
             int j = 0; // ShreddedPaper
             for (int i = 0; i < maxSendsThisTick; ++i) {
                 final long pendingSend = this.sendQueue.firstLong();
                 final int pendingSendX = CoordinateUtils.getChunkX(pendingSend);
                 final int pendingSendZ = CoordinateUtils.getChunkZ(pendingSend);
-                final LevelChunk chunk = ((ChunkSystemLevel)this.world).moonrise$getFullChunkIfLoaded(pendingSendX, pendingSendZ);
-                if (!this.areNeighboursGenerated(pendingSendX, pendingSendZ, 1)) { // ShreddedPaper - do thread check later
+                final LevelChunk chunk = ((ChunkSystemLevel) this.world).moonrise$getFullChunkIfLoaded(pendingSendX, pendingSendZ);
+                if (chunk == null) { // ShreddedPaper - do thread check later / chunk could have been unloaded before sending
                     // nothing to do
                     // the target chunk may not be owned by this region, but this should be resolved in the future
-                    break;
+                    this.retryLaterSendQueue.add(this.sendQueue.dequeueLong());
+                    continue;
                 }
-                if (!((ChunkSystemLevelChunk)chunk).moonrise$isPostProcessingDone()) {
+                if (!chunk.isPostProcessingDone) {
                     // ShreddedPaper start - run post-processing on the tick thread
                     if (!ShreddedPaperChunkTicker.isCurrentlyTickingRegion(this.world, RegionPos.forChunk(chunk.getPos()))) {
-                        this.retryLaterSendQueue.add(this.sendQueue.dequeueLong());
-                        ShreddedPaper.runSync(this.world, chunk.getPos(), () -> {
-                            final LevelChunk chunk2 = this.world.chunkSource.getChunkAtIfLoadedMainThreadNoCache(pendingSendX, pendingSendZ);
-                            if (chunk2 != null && chunk2.areNeighboursLoaded(1) && !chunk2.isPostProcessingDone) {
-                                chunk2.postProcessGeneration();
+                        // Try acquiring a read-only lock before deferring
+                        if (!this.world.chunkScheduler.getRegionLocker().tryReadOnlyLockNow(
+                                RegionPos.forChunk(chunk.getPos()),
+                                () -> this.sendChunk(pendingSendX, pendingSendZ))) {
+                            if (this.retryLaterSendQueue.size() < 1024) {
+                                this.retryLaterSendQueue.add(pendingSend);
                             }
-                        });
-                        if (maxSendsThisTick - i - 1 > this.sendQueue.size() && j++ < maxSendsThisTick * 2)
-                            i--; // Let us try sending more chunks, up to a threshold
-                        continue;
+                            continue;
+                        }
+                    } else {
+                        this.sendChunk(pendingSendX, pendingSendZ);
                     }
+
                     // ShreddedPaper end - run post-processing on the tick thread
+
                     // not yet post-processed, need to do this so that tile entities can properly be sent to clients
                     chunk.postProcessGeneration();
                     // check if there was any recursive action
+                    if (this.removed || this.sendQueue.isEmpty() || this.sendQueue.firstLong() != pendingSend) {
+                        return;
+                    } // else: good to dequeue and send, fall through
                 }
 
+                this.sendQueue.dequeueLong();
+
                 // ShreddedPaper start - send on both player and tick thread
                 if (TickThread.isTickThreadFor(this.world, chunk.getPos())) {
+                    // Add before sendChunk call
                     this.sendChunk(pendingSendX, pendingSendZ);
                 } else if (!this.world.chunkScheduler.getRegionLocker().tryReadOnlyLockNow(RegionPos.forChunk(chunk.getPos()), () -> {
                     this.sendChunk(pendingSendX, pendingSendZ);
                 })) {
                     this.retryLaterSendQueue.add(pendingSend);
-                    if (maxSendsThisTick - i - 1 > this.sendQueue.size() && j++ < maxSendsThisTick * 2) i--; // Let us try sending more chunks, up to a threshold
+                    if (maxSendsThisTick - i - 1 > this.sendQueue.size() && j++ < maxSendsThisTick * 2)
+                        i--; // Let us try sending more chunks, up to a threshold
+                    // Add before each continue
                     continue;
                 }
                 // ShreddedPaper end - send on both player and tick thread
+
                 if (this.removed) {
                     // sendChunk may invoke plugin logic
                     return;
                 }
             }
+
             // ShreddedPaper start - requeue retry send chunks
             if (!this.retryLaterSendQueue.isEmpty()) {
-                for (final long chunk : this.retryLaterSendQueue) {
-                    this.sendQueue.enqueue(chunk);
+                for (final long chunkPos : this.retryLaterSendQueue) {
+                    this.sendQueue.enqueue(chunkPos);
                 }
                 this.retryLaterSendQueue.clear();
             }
             // ShreddedPaper end - requeue retry send chunks
+
             this.flushDelayedTicketOps();
         }
 
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 550afd868eab1914cb15c581ba787cc0e03ff75f..c1715e25ed04f984ae08caa4ee8c1a25e728c03c 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -887,44 +887,50 @@ public class LevelChunk extends ChunkAccess implements ca.spottedleaf.moonrise.p
 
     public boolean isPostProcessingDone; // Paper - replace chunk loader system
     public void postProcessGeneration() {
-        ChunkPos chunkcoordintpair = this.getPos();
+        try {
+            ChunkPos chunkcoordintpair = this.getPos();
 
-        for (int i = 0; i < this.postProcessing.length; ++i) {
-            if (this.postProcessing[i] != null) {
-                ShortListIterator shortlistiterator = this.postProcessing[i].iterator();
+            for (int i = 0; i < this.postProcessing.length; ++i) {
+                if (this.postProcessing[i] != null) {
+                    ShortListIterator shortlistiterator = this.postProcessing[i].iterator();
 
-                while (shortlistiterator.hasNext()) {
-                    Short oshort = (Short) shortlistiterator.next();
-                    BlockPos blockposition = ProtoChunk.unpackOffsetCoordinates(oshort, this.getSectionYFromSectionIndex(i), chunkcoordintpair);
-                    BlockState iblockdata = this.getBlockState(blockposition);
-                    FluidState fluid = iblockdata.getFluidState();
+                    while (shortlistiterator.hasNext()) {
+                        Short oshort = (Short) shortlistiterator.next();
+                        BlockPos blockposition = ProtoChunk.unpackOffsetCoordinates(oshort, this.getSectionYFromSectionIndex(i), chunkcoordintpair);
+                        BlockState iblockdata = this.getBlockState(blockposition);
+                        FluidState fluid = iblockdata.getFluidState();
 
-                    if (!fluid.isEmpty()) {
-                        fluid.tick(this.level, blockposition);
-                    }
+                        if (!fluid.isEmpty()) {
+                            fluid.tick(this.level, blockposition);
+                        }
 
-                    if (!(iblockdata.getBlock() instanceof LiquidBlock)) {
-                        BlockState iblockdata1 = Block.updateFromNeighbourShapes(iblockdata, this.level, blockposition);
+                        if (!(iblockdata.getBlock() instanceof LiquidBlock)) {
+                            BlockState iblockdata1 = Block.updateFromNeighbourShapes(iblockdata, this.level, blockposition);
 
-                        this.level.setBlock(blockposition, iblockdata1, 20);
+                            this.level.setBlock(blockposition, iblockdata1, 20);
+                            if (iblockdata1 != iblockdata)
+                                this.level.chunkSource.blockChanged(blockposition); // Paper - replace player chunk loader - notify since we send before processing full updates
+                        }
                     }
-                }
 
-                this.postProcessing[i].clear();
+                    this.postProcessing[i].clear();
+                }
             }
-        }
 
-        UnmodifiableIterator unmodifiableiterator = ImmutableList.copyOf(this.pendingBlockEntities.keySet()).iterator();
+            UnmodifiableIterator unmodifiableiterator = ImmutableList.copyOf(this.pendingBlockEntities.keySet()).iterator();
 
-        while (unmodifiableiterator.hasNext()) {
-            BlockPos blockposition1 = (BlockPos) unmodifiableiterator.next();
+            while (unmodifiableiterator.hasNext()) {
+                BlockPos blockposition1 = (BlockPos) unmodifiableiterator.next();
 
-            this.getBlockEntity(blockposition1);
-        }
+                this.getBlockEntity(blockposition1);
+            }
 
-        this.pendingBlockEntities.clear();
-        this.upgradeData.upgrade(this);
-        this.postProcessingDone = true; // Paper - rewrite chunk system
+            this.pendingBlockEntities.clear();
+            this.upgradeData.upgrade(this);
+        } finally {
+            this.postProcessingDone = true; // Paper - rewrite chunk system
+            this.isPostProcessingDone = true;
+        }
     }
 
     @Nullable
