--- a/net/minecraft/world/level/ChunkPos.java
+++ b/net/minecraft/world/level/ChunkPos.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.level;
 
 import com.mojang.serialization.Codec;
+import io.multipaper.shreddedpaper.region.RegionPos;
 import io.netty.buffer.ByteBuf;
 import java.util.Spliterators.AbstractSpliterator;
 import java.util.function.Consumer;
@@ -18,14 +_,18 @@
 import org.jspecify.annotations.Nullable;
 
 // Paper start
-public record ChunkPos(int x, int z, long longKey) {
+// ShreddedPaper start - turn back to class
+public class ChunkPos {
+    @Nullable private RegionPos regionPos; // ShreddedPaper
+    public final int x; public int x() { return x; }
+    public final int z; public int z() { return z; }
+    public final long longKey; public long longKey() { return longKey; }
     public ChunkPos(int x, int z) {
-        this(x, z, pack(x, z));
-    }
-
-    public ChunkPos {
-        longKey = pack(x, z); // override if people mess around with longKey
-    }
+        this.x = x;
+        this.z = z;
+        this.longKey = pack(x, z);
+    }
+    // ShreddedPaper end - turn back to class
     // Paper end
     public static final Codec<ChunkPos> CODEC = Codec.INT_STREAM
         .<ChunkPos>comapFlatMap(input -> Util.fixedSize(input, 2).map(ints -> new ChunkPos(ints[0], ints[1])), pos -> IntStream.of(pos.x, pos.z))
@@ -63,6 +_,16 @@
     public static ChunkPos unpack(final long key) {
         return new ChunkPos((int)key, (int)(key >> 32));
     }
+
+    // ShreddedPaper start
+    public RegionPos getRegionPos() {
+        if (this.regionPos == null) {
+            // Cache the RegionPos to minimize object creation
+            this.regionPos = RegionPos.forChunk(this.x, this.z);
+        }
+        return this.regionPos;
+    }
+    // ShreddedPaper end
 
     public static ChunkPos minFromRegion(final int regionX, final int regionZ) {
         return new ChunkPos(regionX << 5, regionZ << 5);
