--- a/net/minecraft/world/level/PotentialCalculator.java
+++ b/net/minecraft/world/level/PotentialCalculator.java
@@ -9,7 +_,7 @@
 
     public void addCharge(final BlockPos pos, final double charge) {
         if (charge != 0.0) {
-            this.charges.add(new PotentialCalculator.PointCharge(pos, charge));
+            this.charges.add(new PotentialCalculator.PointCharge(pos, charge)); // ShreddedPaper - doesn't need a lock, if a value isn't added it's not a big deal
         }
     }
 
@@ -19,8 +_,8 @@
         } else {
             double potentialChange = 0.0;
 
-            for (PotentialCalculator.PointCharge point : this.charges) {
-                potentialChange += point.getPotentialChange(pos);
+            for (int i = 0; i < this.charges.size(); i++) { // ShreddedPaper - prevent CME - list is only ever added to
+                potentialChange += this.charges.get(i).getPotentialChange(pos); // ShreddedPaper - prevent CME - list is only ever added to
             }
 
             return potentialChange * charge;
