--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -74,6 +_,7 @@
     }
 
     public void update(final Level level, final Entity player, final MapItemSavedData data) {
+        synchronized (data) { // ShreddedPaper - thread-safe
         if (level.dimension() == data.dimension && player instanceof Player) {
             int scale = 1 << data.scale;
             int centerX = data.centerX;
@@ -188,6 +_,7 @@
                 }
             }
         }
+        } // ShreddedPaper - thread-safe
     }
 
     private BlockState getCorrectStateForFluidBlock(final Level level, final BlockState state, final BlockPos pos) {
@@ -202,6 +_,7 @@
     public static void renderBiomePreviewMap(final ServerLevel level, final ItemStack mapItemStack) {
         MapItemSavedData data = getSavedData(mapItemStack, level);
         if (data != null) {
+            synchronized (data) { // ShreddedPaper - thread-safe
             data.isExplorerMap = true; // Purpur - Explorer Map API
             if (level.dimension() == data.dimension) {
                 int scale = 1 << data.scale;
@@ -272,6 +_,7 @@
                     }
                 }
             }
+            } // ShreddedPaper - thread-safe
         }
     }
 
