--- a/net/minecraft/world/item/SignItem.java
+++ b/net/minecraft/world/item/SignItem.java
@@ -11,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 
 public class SignItem extends StandingAndWallBlockItem {
-    public static BlockPos openSign; // CraftBukkit
+    public static final ThreadLocal<BlockPos> openSignThreadLocal = new ThreadLocal<>(); // CraftBukkit // ShreddedPaper - thread local for openSign
     public SignItem(final Block sign, final Block wallSign, final Item.Properties properties) {
         super(sign, wallSign, Direction.DOWN, properties);
     }
@@ -32,7 +_,7 @@
             && level.getBlockState(pos).getBlock() instanceof SignBlock sign) {
             // CraftBukkit start - SPIGOT-4678
             // sign.openTextEdit(player, signEntity, true);
-            SignItem.openSign = pos;
+            SignItem.openSignThreadLocal.set(pos); // ShreddedPaper - thread local for openSign
             // CraftBukkit end
         }
 
