--- a/net/minecraft/world/entity/ai/village/VillageSiege.java
+++ b/net/minecraft/world/entity/ai/village/VillageSiege.java
@@ -2,6 +_,7 @@
 
 import com.mojang.logging.LogUtils;
 import java.util.Optional;
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.server.level.ServerLevel;
@@ -96,10 +_,11 @@
     private void trySpawn(final ServerLevel level) {
         Vec3 spawnPos = this.findRandomSpawnPos(level, new BlockPos(this.spawnX, this.spawnY, this.spawnZ));
         if (spawnPos != null) {
+            ShreddedPaper.runSync(level, new BlockPos((int) spawnPos.x, (int) spawnPos.y, (int) spawnPos.z), () -> { // ShreddedPaper - run on block's thread
             Zombie zombie;
             try {
                 zombie = new Zombie(level);
-                zombie.finalizeSpawn(level, level.getCurrentDifficultyAt(zombie.blockPosition()), EntitySpawnReason.EVENT, null);
+                zombie.finalizeSpawn(level, level.getCurrentDifficultyAt(new BlockPos((int) spawnPos.x, (int) spawnPos.y, (int) spawnPos.z)), EntitySpawnReason.EVENT, null); // ShreddedPaper - use zombie's planned position, not 0,0
             } catch (Exception var5) {
                 LOGGER.warn("Failed to create zombie for village siege at {}", spawnPos, var5);
                 com.destroystokyo.paper.exception.ServerInternalException.reportInternalException(var5); // Paper - ServerExceptionEvent
@@ -108,6 +_,7 @@
 
             zombie.snapTo(spawnPos.x, spawnPos.y, spawnPos.z, level.getRandom().nextFloat() * 360.0F, 0.0F);
             level.addFreshEntityWithPassengers(zombie, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.VILLAGE_INVASION); // CraftBukkit
+            }); // ShreddedPaper - run on block's thread
         }
     }
 
