--- a/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
+++ b/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
@@ -2,7 +_,9 @@
 
 import java.util.List;
 import java.util.stream.Collectors;
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.core.BlockPos;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.ContainerUser;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
@@ -136,6 +_,6 @@
     }
 
     private static void scheduleRecheck(Level level, BlockPos pos, BlockState state) {
-        level.scheduleTick(pos, state.getBlock(), 5);
+        ShreddedPaper.ensureSync((ServerLevel) level, pos, () -> level.scheduleTick(pos, state.getBlock(), 5)); // ShreddedPaper - run on container's thread
     }
 }
