--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -509,8 +_,8 @@
             ItemEntity entity = entityFactory.get();
             entity.setDefaultPickUpDelay();
             // CraftBukkit start
-            if (level.captureDrops != null) {
-                level.captureDrops.add(entity);
+            if (level.captureDropsThreadLocal.get() != null) { // ShreddedPaper - use thread local
+                level.captureDropsThreadLocal.get().add(entity); // ShreddedPaper - use thread local
             } else {
                 level.addFreshEntity(entity);
             }
