--- a/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -4,6 +_,9 @@
 import com.mojang.serialization.MapCodec;
 import java.util.Map;
 import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.particles.ParticleTypes;
@@ -142,19 +_,19 @@
     }
 
     @Override
-    public @Nullable TeleportTransition getPortalDestination(final ServerLevel currentLevel, final Entity entity, final BlockPos portalEntryPos) {
+    public CompletableFuture<@Nullable TeleportTransition> getPortalDestination(final ServerLevel currentLevel, final Entity entity, final BlockPos portalEntryPos) { // ShreddedPaper - async teleportation
         ResourceKey<Level> newDimension = currentLevel.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.NETHER ? Level.OVERWORLD : Level.NETHER; // CraftBukkit
         ServerLevel newLevel = currentLevel.getServer().getLevel(newDimension);
         // Paper start - Add EntityPortalReadyEvent
         io.papermc.paper.event.entity.EntityPortalReadyEvent portalReadyEvent = new io.papermc.paper.event.entity.EntityPortalReadyEvent(entity.getBukkitEntity(), newLevel == null ? null : newLevel.getWorld(), org.bukkit.PortalType.NETHER);
         if (!portalReadyEvent.callEvent()) {
             entity.portalProcess = null;
-            return null;
+            return CompletableFuture.completedFuture(null); // ShreddedPaper - async teleportation
         }
         newLevel = portalReadyEvent.getTargetWorld() == null ? null : ((org.bukkit.craftbukkit.CraftWorld) portalReadyEvent.getTargetWorld()).getHandle();
         // Paper end - Add EntityPortalReadyEvent
         if (newLevel == null) {
-            return null;
+            return CompletableFuture.completedFuture(null); // ShreddedPaper - async teleportation
         } else {
             boolean toNether = newLevel.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.NETHER; // CraftBukkit
             WorldBorder newWorldBorder = newLevel.getWorldBorder();
@@ -169,17 +_,17 @@
             // CraftBukkit start
             org.bukkit.craftbukkit.event.PortalEventResult result = org.bukkit.craftbukkit.event.CraftEventFactory.handlePortalEvents(entity, org.bukkit.craftbukkit.util.CraftLocation.toBukkit(approximateExitPos, newLevel), org.bukkit.PortalType.NETHER, portalSearchRadius, newLevel.paperConfig().environment.portalCreateRadius); // Paper - use custom portal search radius
             if (result == null) {
-                return null;
+                return CompletableFuture.completedFuture(null); // ShreddedPaper - async teleportation
             }
             newLevel = ((org.bukkit.craftbukkit.CraftWorld) result.to().getWorld()).getHandle();
             newWorldBorder = newLevel.getWorldBorder();
             approximateExitPos = newWorldBorder.clampToBounds(result.to().getX(), result.to().getY(), result.to().getZ());
-            return this.getExitPortal(newLevel, entity, portalEntryPos, approximateExitPos, toNether, newWorldBorder, result);
+            return this.getExitPortalAsync(newLevel, entity, portalEntryPos, approximateExitPos, toNether, newWorldBorder, result);  // ShreddedPaper - async teleportation
             // CraftBukkit end
         }
     }
 
-    private @Nullable TeleportTransition getExitPortal(
+    private CompletableFuture<@Nullable TeleportTransition> getExitPortalAsync( // ShreddedPaper - async teleportation
         final ServerLevel newLevel,
         final Entity entity,
         final BlockPos portalEntryPos,
@@ -188,6 +_,7 @@
         final WorldBorder worldBorder
         , org.bukkit.craftbukkit.event.PortalEventResult result // CraftBukkit
     ) {
+        return CompletableFuture.supplyAsync(() -> { // ShreddedPaper - async teleportation
         Optional<BlockPos> exitPortalPos = newLevel.getPortalForcer().findClosestPortalPosition(approximateExitPos, worldBorder, result.searchRadius()); // CraftBukkit
         BlockUtil.FoundRectangle exitPortal;
         TeleportTransition.PostTeleportTransition post;
@@ -221,6 +_,7 @@
         // CraftBukkit end
 
         return getDimensionTransitionFromExit(entity, portalEntryPos, exitPortal, newLevel, post);
+        }, r -> ShreddedPaper.ensureSync(newLevel, approximateExitPos, r)); // ShreddedPaper - async teleportation
     }
 
     private static TeleportTransition getDimensionTransitionFromExit(
