--- a/net/minecraft/world/phys/Vec3.java
+++ b/net/minecraft/world/phys/Vec3.java
@@ -1,9 +_,11 @@
 package net.minecraft.world.phys;
 
 import com.mojang.serialization.Codec;
+import io.multipaper.shreddedpaper.region.RegionPos;
 import io.netty.buffer.ByteBuf;
 import java.util.EnumSet;
 import java.util.List;
+
 import net.minecraft.core.Direction;
 import net.minecraft.core.Position;
 import net.minecraft.core.Vec3i;
@@ -338,4 +_,14 @@
     public boolean isFinite() {
         return Double.isFinite(this.x) && Double.isFinite(this.y) && Double.isFinite(this.z);
     }
+
+    // ShreddedPaper start - cap velocity to region size
+    public Vec3 shreddedPaperCapToRegionSize() {
+        double horizontalDistanceSqr = this.horizontalDistanceSqr();
+        if (horizontalDistanceSqr > RegionPos.MAX_DISTANCE_SQR) {
+            return this.scale((RegionPos.MAX_DISTANCE - 1) / Math.sqrt(horizontalDistanceSqr)); // Subtract 1 to ensure no double precision issues causes horizontalDistance > RegionPos.MAX_DISTANCE
+        }
+        return this;
+    }
+    // ShreddedPaper end - cap velocity to region size
 }
