--- a/net/minecraft/world/entity/raid/Raids.java
+++ b/net/minecraft/world/entity/raid/Raids.java
@@ -1,12 +_,15 @@
 package net.minecraft.world.entity.raid;
 
+import com.google.common.collect.Maps;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
 import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap.Entry;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
 import java.util.OptionalInt;
 import net.minecraft.core.BlockPos;
 import net.minecraft.nbt.CompoundTag;
@@ -35,14 +_,14 @@
                 Raids.RaidWithId.CODEC
                     .listOf()
                     .optionalFieldOf("raids", List.of())
-                    .forGetter(r -> r.raidMap.int2ObjectEntrySet().stream().map(Raids.RaidWithId::from).toList()),
+                    .forGetter(r -> r.raidMap.entrySet().stream().map(Raids.RaidWithId::from).toList()), // ShreddedPaper - concurrent map
                 Codec.INT.fieldOf("next_id").forGetter(r -> r.nextId),
                 Codec.INT.fieldOf("tick").forGetter(r -> r.tick)
             )
             .apply(i, Raids::new)
     );
     public static final SavedDataType<Raids> TYPE = new SavedDataType<>(RAID_FILE_ID, Raids::new, CODEC, DataFixTypes.SAVED_DATA_RAIDS);
-    public final Int2ObjectMap<Raid> raidMap = new Int2ObjectOpenHashMap<>();
+    public final Map<Integer, Raid> raidMap = Maps.newConcurrentMap(); // ShreddedPaper - concurrent map
     private int nextId = 1;
     private int tick;
 
@@ -65,9 +_,9 @@
     }
 
     public OptionalInt getId(final Raid raid) {
-        for (Entry<Raid> entry : this.raidMap.int2ObjectEntrySet()) {
+        for (Map.Entry<Integer, Raid> entry : this.raidMap.entrySet()) { // ShreddedPaper - concurrent map
             if (entry.getValue() == raid) {
-                return OptionalInt.of(entry.getIntKey());
+                return OptionalInt.of(entry.getKey());
             }
         }
 
@@ -91,16 +_,18 @@
 
         while (raidIterator.hasNext()) {
             Raid raid = raidIterator.next();
+            ShreddedPaper.runSync(level, raid.getCenter(), () -> { // ShreddedPaper - run on right thread
             if (!level.getGameRules().get(GameRules.RAIDS)) {
                 raid.stop();
             }
 
             if (raid.isStopped()) {
-                raidIterator.remove();
+                this.raidMap.remove(raid.idOrNegativeOne, raid); // raidIterator.remove(); // ShreddedPaper - run on right thread
                 this.setDirty();
             } else {
                 raid.tick(level);
             }
+            }); // ShreddedPaper - run on right thread
         }
 
         if (this.tick % 200 == 0) {
@@ -208,8 +_,8 @@
                 .apply(i, Raids.RaidWithId::new)
         );
 
-        public static Raids.RaidWithId from(final Entry<Raid> entry) {
-            return new Raids.RaidWithId(entry.getIntKey(), entry.getValue());
+        public static Raids.RaidWithId from(final Map.Entry<Integer, Raid> entry) { // ShreddedPaper - concurrent map
+            return new Raids.RaidWithId(entry.getKey(), entry.getValue()); // ShreddedPaper - concurrent map
         }
     }
 }
