--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -15,6 +_,7 @@
 import java.util.UUID;
 import java.util.function.Predicate;
 import java.util.stream.Stream;
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.SharedConstants;
 import net.minecraft.advancements.triggers.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
@@ -423,12 +_,15 @@
                         for (UUID heroUUID : this.heroesOfTheVillage) {
                             Entity entity = level.getEntity(heroUUID);
                             if (entity instanceof LivingEntity hero && !entity.isSpectator()) {
+                                ShreddedPaper.ensureSync(hero, () -> { // ShreddedPaper - run on right thread
                                 hero.addEffect(new MobEffectInstance(MobEffects.HERO_OF_THE_VILLAGE, 48000, this.raidOmenLevel - 1, false, false, true));
                                 if (hero instanceof ServerPlayer playerHero) {
                                     playerHero.awardStat(Stats.RAID_WIN);
                                     CriteriaTriggers.RAID_WIN.trigger(playerHero);
-                                    winners.add(playerHero.getBukkitEntity()); // CraftBukkit
+                                    // winners.add(playerHero.getBukkitEntity()); // CraftBukkit // ShreddedPaper - this doesn't need to be run on right thread
                                 }
+                                }); // ShreddedPaper - run on right thread
+                                if (hero instanceof ServerPlayer playerHero) winners.add(playerHero.getBukkitEntity()); // CraftBukkit // ShreddedPaper - this doesn't need to be run on right thread
                             }
                         }
                         org.bukkit.craftbukkit.event.CraftEventFactory.callRaidFinishEvent(level, this, winners); // CraftBukkit
