--- a/net/minecraft/world/scores/PlayerTeam.java
+++ b/net/minecraft/world/scores/PlayerTeam.java
@@ -11,6 +_,7 @@
 import java.util.List;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.ComponentSerialization;
@@ -25,7 +_,7 @@
     private static final byte BIT_SEE_INVISIBLES = 2;
     private final Scoreboard scoreboard;
     private final String name;
-    private final Set<String> players = Sets.newHashSet();
+    private final Set<String> players = ConcurrentHashMap.newKeySet(); // ShreddedPaper - concurrent
     private Component displayName;
     private Component playerPrefix = CommonComponents.EMPTY;
     private Component playerSuffix = CommonComponents.EMPTY;
