--- a/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
+++ b/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
@@ -13,6 +_,7 @@
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Predicate;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
@@ -69,6 +_,7 @@
     public byte scale;
     public byte[] colors = new byte[16384];
     public boolean locked;
+    public AtomicBoolean hasCalledMapInitializeEvent = new AtomicBoolean(false); // ShreddedPaper
     private final org.bukkit.craftbukkit.map.RenderData vanillaRender = new org.bukkit.craftbukkit.map.RenderData(); // Paper - Use Vanilla map renderer when possible
     public final List<MapItemSavedData.HoldingPlayer> carriedBy = Lists.newArrayList();
     public final Map<Player, MapItemSavedData.HoldingPlayer> carriedByPlayers = Maps.newHashMap();
@@ -181,7 +_,7 @@
         return new MapItemSavedData(0, 0, scale, false, false, isLocked, dimension);
     }
 
-    public MapItemSavedData locked() {
+    public synchronized MapItemSavedData locked() { // ShreddedPaper - thread-safe
         MapItemSavedData result = new MapItemSavedData(
             this.centerX, this.centerZ, this.scale, this.trackingPosition, this.unlimitedTracking, true, this.dimension
         );
@@ -201,7 +_,7 @@
         return stack -> stack == mapStack || stack.is(mapStack.getItem()) && Objects.equals(mapId, stack.get(DataComponents.MAP_ID));
     }
 
-    public void tickCarriedBy(final Player tickingPlayer, final ItemStack itemStack, final @Nullable ItemFrame placedInFrame) {
+    public synchronized void tickCarriedBy(final Player tickingPlayer, final ItemStack itemStack, final @Nullable ItemFrame placedInFrame) { // ShreddedPaper - thread-safe
         if (!this.carriedByPlayers.containsKey(tickingPlayer)) {
             MapItemSavedData.HoldingPlayer holdingPlayer = new MapItemSavedData.HoldingPlayer(tickingPlayer);
             this.carriedByPlayers.put(tickingPlayer, holdingPlayer);
@@ -404,7 +_,7 @@
         }
     }
 
-    public @Nullable Packet<?> getUpdatePacket(final MapId id, final Player player) {
+    public synchronized @Nullable Packet<?> getUpdatePacket(final MapId id, final Player player) { // ShreddedPaper - thread-safe
         MapItemSavedData.HoldingPlayer holdingPlayer = this.carriedByPlayers.get(player);
         return holdingPlayer == null ? null : holdingPlayer.nextUpdatePacket(id);
     }
@@ -518,7 +_,7 @@
         return false;
     }
 
-    public void addClientSideDecorations(final List<MapDecoration> decorations) {
+    public synchronized void addClientSideDecorations(final List<MapDecoration> decorations) { // ShreddedPaper - thread-safe
         this.decorations.clear();
         this.trackedDecorationCount = 0;
 
