From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: caleb <graham.caleb.m@gmail.com>
Date: Fri, 28 Nov 2025 11:25:31 -0600
Subject: [PATCH] Level-ticks


diff --git a/io/multipaper/shreddedpaper/region/LevelTicksRegionProxy.java b/io/multipaper/shreddedpaper/region/LevelTicksRegionProxy.java
new file mode 100644
index 0000000000000000000000000000000000000000..480e9ff2406ccf8769f9819ffa6b94a0dd942489
--- /dev/null
+++ b/io/multipaper/shreddedpaper/region/LevelTicksRegionProxy.java
@@ -0,0 +1,111 @@
+package io.multipaper.shreddedpaper.region;
+
+import com.mojang.logging.LogUtils;
+import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Vec3i;
+import net.minecraft.util.profiling.ProfilerFiller;
+import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.levelgen.structure.BoundingBox;
+import net.minecraft.world.ticks.LevelChunkTicks;
+import net.minecraft.world.ticks.LevelTicks;
+import net.minecraft.world.ticks.ScheduledTick;
+import org.slf4j.Logger;
+import io.multipaper.shreddedpaper.util.SimpleStampedLock;
+
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import java.util.function.LongPredicate;
+import java.util.function.Supplier;
+
+public class LevelTicksRegionProxy<T> extends LevelTicks<T> {
+    private static final Logger LOGGER = LogUtils.getClassLogger();
+    private final LongPredicate tickingFutureReadyPredicate;
+    private final Supplier<ProfilerFiller> profilerGetter;
+    private final Long2ObjectMap<LevelTicks<T>> regions = new Long2ObjectOpenHashMap<>();
+    private final SimpleStampedLock regionsLock = new SimpleStampedLock();
+
+    public LevelTicksRegionProxy(LongPredicate tickingFutureReadyPredicate, Supplier<ProfilerFiller> profilerGetter) {
+        super(tickingFutureReadyPredicate, profilerGetter);
+        this.tickingFutureReadyPredicate = tickingFutureReadyPredicate;
+        this.profilerGetter = profilerGetter;
+    }
+
+    private LevelTicks<T> createRegionLevelTicks() {
+        return new LevelTicks<>(tickingFutureReadyPredicate, profilerGetter);
+    }
+
+    public Optional<LevelTicks<T>> get(BlockPos pos) {
+        return get(new ChunkPos(pos));
+    }
+
+    public Optional<LevelTicks<T>> get(ChunkPos pos) {
+        return get(RegionPos.forChunk(pos));
+    }
+
+    public Optional<LevelTicks<T>> get(RegionPos pos) {
+        return Optional.ofNullable(regionsLock.optimisticRead(() -> regions.get(pos.longKey)));
+    }
+
+    public void addContainer(ChunkPos pos, LevelChunkTicks<T> scheduler) {
+        get(pos).orElseGet(() -> {
+            return regionsLock.write(() -> regions.computeIfAbsent(RegionPos.forChunk(pos).longKey, k -> createRegionLevelTicks()));
+        }).addContainer(pos, scheduler);
+    }
+
+    public void removeContainer(ChunkPos pos) {
+        get(pos).ifPresent(v -> {
+            v.removeContainer(pos);
+            if (v.isEmpty()) {
+                regionsLock.write(() -> regions.remove(RegionPos.forChunk(pos).longKey));
+            }
+        });
+    }
+
+    @Override
+    public void schedule(ScheduledTick<T> orderedTick) {
+        get(orderedTick.pos()).orElseThrow(() -> new IllegalArgumentException("Chunk not loaded: " + orderedTick.pos())).schedule(orderedTick);
+    }
+
+    public void tick(RegionPos regionPos, long time, int maxTicks, BiConsumer<BlockPos, T> ticker) {
+        get(regionPos).ifPresent(v -> {
+            v.tick(time, maxTicks, ticker);
+        });
+    }
+
+    @Override
+    public void tick(long time, int maxTicks, BiConsumer<BlockPos, T> ticker) {
+        // Do nothing
+    }
+
+    @Override
+    public boolean hasScheduledTick(BlockPos pos, T type) {
+        return get(pos).map(v -> v.hasScheduledTick(pos, type)).orElse(false);
+    }
+
+    @Override
+    public boolean willTickThisTick(BlockPos pos, T type) {
+        return get(pos).map(v -> v.willTickThisTick(pos, type)).orElse(false);
+    }
+
+    @Override
+    public void clearArea(BoundingBox box) {
+        // Surely no one will miss this
+    }
+
+    @Override
+    public void copyArea(BoundingBox box, Vec3i offset) {
+        // Surely no one will miss this
+    }
+
+    @Override
+    public void copyAreaFrom(LevelTicks<T> scheduler, BoundingBox box, Vec3i offset) {
+        // Surely no one will miss this
+    }
+
+    @Override
+    public int count() {
+        return -1;
+    }
+}
\ No newline at end of file
diff --git a/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java b/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
index 2a1eb3d01ef53c552b748e6d2b9b1b1d407c1b7c..beb4c7d37fee33b2670efb6715aae5b6213a6576 100644
--- a/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
+++ b/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
@@ -60,6 +60,10 @@ public class ShreddedPaperChunkTicker {
             });
 
             region.tickTasks();
+
+            level.blockTicks.tick(region.getRegionPos(), level.getGameTime(), level.paperConfig().environment.maxBlockTicks, level::tickBlock);
+            level.fluidTicks.tick(region.getRegionPos(), level.getGameTime(), level.paperConfig().environment.maxBlockTicks, level::tickFluid);
+
             region.forEachTickingChunk(chunk -> _tickChunk(level, chunk, spawnercreature_d));
 
             region.forEachTickingEntity(ShreddedPaperEntityTicker::tickEntity);
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 5b32749d88f07e4c5113bbb4599bddfe86ffb865..cdbbd49d9dc8b23c9fe5128bc55a80acd71aa1d7 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1,5 +1,6 @@
 package net.minecraft.server.level;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.Lists;
 import com.mojang.datafixers.DataFixer;
@@ -165,6 +166,7 @@ import net.minecraft.world.phys.shapes.VoxelShape;
 import net.minecraft.world.ticks.LevelTicks;
 import net.minecraft.world.waypoints.WaypointTransmitter;
 import org.slf4j.Logger;
+import io.multipaper.shreddedpaper.region.LevelTicksRegionProxy;
 import io.multipaper.shreddedpaper.region.RegionPos;
 import io.multipaper.shreddedpaper.threading.ShreddedPaperRegionScheduler;
 
@@ -190,8 +192,8 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     private final SleepStatus sleepStatus;
     private int emptyTime;
     private final PortalForcer portalForcer;
-    private final LevelTicks<Block> blockTicks = new LevelTicks<>(this::isPositionTickingWithEntitiesLoaded);
-    private final LevelTicks<Fluid> fluidTicks = new LevelTicks<>(this::isPositionTickingWithEntitiesLoaded);
+    public final LevelTicksRegionProxy<Block> blockTicks; // ShreddedPaper
+    public final LevelTicksRegionProxy<Fluid> fluidTicks; // ShreddedPaper
     private final PathTypeCache pathTypesByPosCache = new PathTypeCache();
     final Set<Mob> navigatingMobs = new ObjectOpenHashSet<>();
     volatile boolean isUpdatingNavigations;
@@ -595,6 +597,8 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         this.uuid = org.bukkit.craftbukkit.util.WorldUUID.getOrCreate(levelStorageAccess.levelDirectory.path().toFile());
         this.players = new CopyOnWriteArrayList<>(); // ShreddedPaper - thread-safe (players should not be changing worlds often, thus copy-on-write is sufficient)
         // CraftBukkit end
+        this.blockTicks = new LevelTicksRegionProxy<>(this::isPositionTickingWithEntitiesLoaded, this.getProfilerSupplier()); // ShreddedPaper
+        this.fluidTicks = new LevelTicksRegionProxy<>(this::isPositionTickingWithEntitiesLoaded, this.getProfilerSupplier()); // ShreddedPaper
         this.tickTime = tickTime;
         this.server = server;
         // Purpur start - Allow toggling special MobSpawners per world
@@ -1321,7 +1325,8 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         this.emptyTime = 0;
     }
 
-    private void tickFluid(BlockPos pos, Fluid fluid) {
+    public void tickFluid(BlockPos pos, Fluid fluid) { // ShreddedPaper - make public
+        TickThread.ensureTickThread(this, pos, "Cannot tick fluid outside of tick thread"); // ShreddedPaper
         BlockState blockState = this.getBlockState(pos);
         FluidState fluidState = blockState.getFluidState();
         if (fluidState.is(fluid)) {
@@ -1329,20 +1334,21 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         }
         // Paper start - rewrite chunk system
         if ((++this.tickedBlocksOrFluids & 7L) != 0L) {
-            ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks();
+            // ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks(); // ShreddedPaper - unnecessary now
         }
         // Paper end - rewrite chunk system
 
     }
 
-    private void tickBlock(BlockPos pos, Block block) {
+    public void tickBlock(BlockPos pos, Block block) {
+        TickThread.ensureTickThread(this, pos, "Cannot tick block outside of tick thread"); // ShreddedPaper
         BlockState blockState = this.getBlockState(pos);
         if (blockState.is(block)) {
             blockState.tick(this, pos, this.random);
         }
         // Paper start - rewrite chunk system
         if ((++this.tickedBlocksOrFluids & 7L) != 0L) {
-            ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks();
+            // ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks(); // ShreddedPaper - unnecessary now
         }
         // Paper end - rewrite chunk system
 
diff --git a/net/minecraft/world/level/ScheduledTickAccess.java b/net/minecraft/world/level/ScheduledTickAccess.java
index 0fbbccde3fb2143fba0933a5fcb6dc4c0bcc9501..8e84d0d4d7317b3b7594ee6bcc48da56428e249e 100644
--- a/net/minecraft/world/level/ScheduledTickAccess.java
+++ b/net/minecraft/world/level/ScheduledTickAccess.java
@@ -1,6 +1,8 @@
 package net.minecraft.world.level;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import net.minecraft.core.BlockPos;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.ticks.LevelTickAccess;
@@ -15,20 +17,24 @@ public interface ScheduledTickAccess {
     LevelTickAccess<Block> getBlockTicks();
 
     default void scheduleTick(BlockPos pos, Block block, int delay, TickPriority priority) {
+        if (this instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, pos, "Cannot schedule outside of tick thread"); // ShreddedPaper
         this.getBlockTicks().schedule(this.createTick(pos, block, delay, priority));
     }
 
     default void scheduleTick(BlockPos pos, Block block, int delay) {
+        if (this instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, pos, "Cannot schedule outside of tick thread"); // ShreddedPaper
         this.getBlockTicks().schedule(this.createTick(pos, block, delay));
     }
 
     LevelTickAccess<Fluid> getFluidTicks();
 
     default void scheduleTick(BlockPos pos, Fluid fluid, int delay, TickPriority priority) {
+        if (this instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, pos, "Cannot schedule outside of tick thread"); // ShreddedPaper
         this.getFluidTicks().schedule(this.createTick(pos, fluid, delay, priority));
     }
 
     default void scheduleTick(BlockPos pos, Fluid fluid, int delay) {
+        if (this instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, pos, "Cannot schedule outside of tick thread"); // ShreddedPaper
         this.getFluidTicks().schedule(this.createTick(pos, fluid, delay));
     }
 }
diff --git a/net/minecraft/world/ticks/LevelTicks.java b/net/minecraft/world/ticks/LevelTicks.java
index 7ea0273018b1a364a5e2b7bf57da840383c8ffa5..e0f89d77549bae8f4290751a4c6011bc168aad15 100644
--- a/net/minecraft/world/ticks/LevelTicks.java
+++ b/net/minecraft/world/ticks/LevelTicks.java
@@ -285,4 +285,10 @@ public class LevelTicks<T> implements LevelTickAccess<T> {
     interface PosAndContainerConsumer<T> {
         void accept(long pos, LevelChunkTicks<T> container);
     }
+
+    // ShreddedPaper start
+    public boolean isEmpty() {
+        return this.allContainers.isEmpty();
+    }
+    // ShreddedPaper end
 }
