From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: caleb <graham.caleb.m@gmail.com>
Date: Fri, 28 Nov 2025 13:00:47 -0600
Subject: [PATCH] EntityLookup-accessibleEntities


diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index f0f5aa48af099c8244cd59da2c1dac6744a41abe..0b0969548e66e11574ac6e07946cd804cddf64cc 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -81,6 +81,7 @@ public final class NaturalSpawner {
             if (!(entity instanceof Mob mob && (mob.isPersistenceRequired() || mob.requiresCustomPersistence()))) {
                 MobCategory category = entity.getType().getCategory();
                 if (category != MobCategory.MISC) {
+                    if (entity == null) continue; // ShreddedPaper - concurrent modification
                     // Paper start - Only count natural spawns
                     if (!entity.level().paperConfig().entities.spawning.countAllMobsForSpawning &&
                         !(entity.spawnReason == org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL ||
