From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: caleb <graham.caleb.m@gmail.com>
Date: Mon, 1 Dec 2025 01:12:30 -0600
Subject: [PATCH] teleportIfOutsideBorder-use-teleportAsync


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index a3b985cc454b2c62e7642f7eb05de590ecde9272..e669f2c0ccc0e2368a8eb6bb8ec858e34d31f376 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -451,7 +451,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                     if (damagePerBlock > 0.0) {
                         // Purpur start - Add option to teleport to spawn if outside world border
                         if (this.level().purpurConfig.teleportIfOutsideBorder && this instanceof ServerPlayer serverPlayer) {
-                            serverPlayer.teleport(org.bukkit.craftbukkit.util.CraftLocation.toBukkit(this.level().getSharedSpawnPos(), this.level()));
+                            serverPlayer.getBukkitEntity().teleportAsync(io.papermc.paper.util.MCUtil.toLocation(level(), ((ServerLevel) level()).getSharedSpawnPos())); return; } // Purpur // ShreddedPaper - use teleportAsync
                             return;
                         }
                         // Purpur end - Add option to teleport to spawn if outside world border
@@ -488,7 +488,6 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 this.lastPos = blockPos;
                 this.onChangedBlock(serverLevel1, blockPos);
             }
-        }
 
         if (this.hurtTime > 0) {
             this.hurtTime--;
