From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: caleb <graham.caleb.m@gmail.com>
Date: Mon, 1 Dec 2025 01:30:31 -0600
Subject: [PATCH] handlingTick


diff --git a/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java b/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
index a65b83f5f5296a947891070b13d098215d1a39aa..72e0ca437f3ef8740332792d2838f8dfd8697df4 100644
--- a/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
+++ b/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
@@ -72,6 +72,8 @@ public class ShreddedPaperChunkTicker {
 
             region.tickTasks();
 
+            level.handlingTickThreadLocal.set(true);
+
             level.blockTicks.tick(region.getRegionPos(), level.getGameTime(), level.paperConfig().environment.maxBlockTicks, level::tickBlock);
             level.fluidTicks.tick(region.getRegionPos(), level.getGameTime(), level.paperConfig().environment.maxBlockTicks, level::tickFluid);
 
@@ -79,6 +81,8 @@ public class ShreddedPaperChunkTicker {
 
             level.runBlockEvents(region);
 
+            level.handlingTickThreadLocal.set(false);
+
             region.forEachTickingEntity(ShreddedPaperEntityTicker::tickEntity);
 
             region.forEachTrackedEntity(ShreddedPaperEntityTicker::processTrackQueue);
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 182ec4b3ffaffea3f8e6dc13789e71d8a6c8eeaf..e1986d50e744d4e814e36631fd9abd96c4c953a7 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -199,7 +199,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     // private final ObjectLinkedOpenHashSet<BlockEventData> blockEvents = new ObjectLinkedOpenHashSet<>(); // ShreddedPaper - moved into each region
     private final ThreadLocal<List<BlockEventData>> blockEventsToRescheduleThreadLocal; // ShreddedPaper
     private final List<BlockEventData> blockEventsToReschedule = new ArrayList<>(64);
-    private boolean handlingTick;
+    public ThreadLocal<Boolean> handlingTickThreadLocal = ThreadLocal.withInitial(() -> false); // ShreddedPaper
     private final List<CustomSpawner> customSpawners;
     @Nullable
     private EndDragonFight dragonFight;
@@ -738,7 +738,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
 
     public void tick(BooleanSupplier hasTimeLeft) {
         ProfilerFiller profilerFiller = Profiler.get();
-        this.handlingTick = true;
+        this.handlingTickThreadLocal.set(true); // ShreddedPaper
         TickRateManager tickRateManager = this.tickRateManager();
         boolean runsNormally = tickRateManager.runsNormally();
         if (runsNormally) {
@@ -802,7 +802,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
             // this.runBlockEvents(); // ShreddedPaper - handled locally in the region
         }
 
-        this.handlingTick = false;
+        this.handlingTickThreadLocal.set(false); // ShreddedPaper
         profilerFiller.pop();
         boolean flag = !paperConfig().unsupportedSettings.disableWorldTickingWhenEmpty || !this.players.isEmpty() || !this.getForceLoadedChunks().isEmpty(); // CraftBukkit - this prevents entity cleanup, other issues on servers with no players // Paper - restore this
         if (flag) {
@@ -1135,7 +1135,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     }
 
     public boolean isHandlingTick() {
-        return this.handlingTick;
+        return this.handlingTickThreadLocal.get();
     }
 
     public boolean canSleepThroughNights() {
