From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: caleb <graham.caleb.m@gmail.com>
Date: Mon, 1 Dec 2025 01:47:19 -0600
Subject: [PATCH] Send-ping-packet-later


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index f07d59fa033c3ab19bffd9a260797d6364c02a21..a7b3ce551a9e7199b136103595840542653a06f2 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1826,6 +1826,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
         for (ServerPlayer serverPlayer : this.playerList.getPlayers()) {
             serverPlayer.connection.chunkSender.sendNextChunks(serverPlayer);
+            serverPlayer.connection.keepConnectionAlive(); // ShreddedPaper - send the ping packet right before flushing
             serverPlayer.connection.resumeFlushing();
         }
 
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index ab1896fe5434cfa1e535d7c41363b2e6fca27967..289c9ed9ac090427dff753acc1aa84f0f8b6d9ce 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -403,7 +403,7 @@ public class ServerGamePacketListenerImpl
             this.aboveGroundVehicleTickCount = 0;
         }
 
-        this.keepConnectionAlive();
+        // this.keepConnectionAlive(); // ShreddedPaper - called elsewhere
         this.chatSpamThrottler.tick();
         this.dropSpamThrottler.tick();
         this.tabSpamThrottler.tick(); // Paper - configurable tab spam limits
