From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: caleb <graham.caleb.m@gmail.com>
Date: Mon, 1 Dec 2025 01:59:53 -0600
Subject: [PATCH] Don-t-allow-actions-outside-of-our-region


diff --git a/ca/spottedleaf/moonrise/common/misc/NearbyPlayers.java b/ca/spottedleaf/moonrise/common/misc/NearbyPlayers.java
index 56bdb25a6bcc414b860039ec045158aade2aaac5..833a223bfdde19c1c2255bf8bc6cbc6469d6055b 100644
--- a/ca/spottedleaf/moonrise/common/misc/NearbyPlayers.java
+++ b/ca/spottedleaf/moonrise/common/misc/NearbyPlayers.java
@@ -81,7 +81,7 @@ public final class NearbyPlayers {
             if (this.players.putIfAbsent(player, newTrackers) != null) {
                 throw new IllegalStateException("Already have player " + player);
             }
-        }; // ShreddedPaper
+        }); // ShreddedPaper
 
         final ChunkPos chunk = player.chunkPosition();
 
@@ -140,7 +140,7 @@ public final class NearbyPlayers {
     }
 
     public TrackedChunk getChunk(final int chunkX, final int chunkZ) {
-        return byChunkLock.optimisticRead(() -> this.byChunk.get(CoordinateUtils.getChunkKey(chunkX,chunkZ)); // ShreddedPaper
+        return byChunkLock.optimisticRead(() -> this.byChunk.get(CoordinateUtils.getChunkKey(chunkX,chunkZ))); // ShreddedPaper
     }
 
     public ReferenceList<ServerPlayer> getPlayers(final BlockPos pos, final NearbyMapType type) {
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index 0accca9c10dbc7e6841def7741c237e98ebfa55f..05fa48b60ea97e12453ecd8be3fe6a9a99fa2a49 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -836,7 +836,7 @@ public final class RegionizedPlayerChunkLoader {
                     break;
                 }
                 if (!((ChunkSystemLevelChunk)chunk).moonrise$isPostProcessingDone()) {
-2                    // ShreddedPaper start - run on region's thread with write lock
+                    // ShreddedPaper start - run on region's thread with write lock
                     Runnable postProcessingRunOnThread = () -> {
                         // not yet post-processed, need to do this so that tile entities can properly be sent to clients
                         chunk.postProcessGeneration();
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
index 2106f2d45c120b391fe0fab3ae54a91fdc1ef1b3..6081defe767762b19a5b7deda65effe162c9b9d7 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
@@ -118,7 +118,8 @@ public final class ChunkHolderManager {
         this.world = world;
         this.taskScheduler = taskScheduler;
         this.ticketLockArea = new ReentrantAreaLock(taskScheduler.getChunkSystemLockShift());
-        // this.unloadQueue = new ChunkUnloadQueue(((ChunkSystemServerLevel)world).moonrise$getRegionChunkShift()); // ShreddedPaper - use our own unload queue    }
+        // this.unloadQueue = new ChunkUnloadQueue(((ChunkSystemServerLevel)world).moonrise$getRegionChunkShift()); // ShreddedPaper - use our own unload queue
+    }
 
     public boolean processTicketUpdates(final int chunkX, final int chunkZ) {
         final int ticketShift = ThreadedTicketLevelPropagator.SECTION_SHIFT;
@@ -1318,9 +1319,7 @@ public final class ChunkHolderManager {
                 } finally {
                     this.taskScheduler.schedulingLockArea.unlock(scheduleLock);
                 }
-            } finally {
-                this.ticketLockArea.unlock(ticketLock);
-            }
+
 
             // stage 2: invoke expensive unload logic, designed to run without locks thanks to stage 1
             final List<NewChunkHolder> stage3 = new ArrayList<>(stage2.size());
diff --git a/net/minecraft/server/level/ServerPlayerGameMode.java b/net/minecraft/server/level/ServerPlayerGameMode.java
index f668b0778e1bce40271284fefd470f10732993d0..cd0f7567dce07fd53af9588b0830b16572d725ea 100644
--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -1,5 +1,6 @@
 package net.minecraft.server.level;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import com.mojang.logging.LogUtils;
 import java.util.Objects;
 import javax.annotation.Nullable;
@@ -115,7 +116,7 @@ public class ServerPlayerGameMode {
         // this.gameTicks = net.minecraft.server.MinecraftServer.currentTick; // CraftBukkit
         this.gameTicks = (int) this.level.getLagCompensationTick(); // Paper - lag compensate eating
         if (this.hasDelayedDestroy) {
-            BlockState blockState = this.level.getBlockStateIfLoaded(this.delayedDestroyPos); // Paper - Don't allow digging into unloaded chunks
+            BlockState blockState = !TickThread.isTickThreadFor(this.level, this.delayedDestroyPos) ? null : this.level.getBlockStateIfLoaded(this.delayedDestroyPos); // Paper - Don't allow digging into unloaded chunks // ShreddedPaper - Don't allow digging into chunks outside our region
             if (blockState == null || blockState.isAir()) { // Paper - Don't allow digging into unloaded chunks
                 this.hasDelayedDestroy = false;
             } else {
@@ -127,7 +128,7 @@ public class ServerPlayerGameMode {
             }
         } else if (this.isDestroyingBlock) {
             // Paper start - Don't allow digging into unloaded chunks; don't want to do same logic as above, return instead
-            BlockState blockState = this.level.getBlockStateIfLoaded(this.destroyPos);
+            BlockState blockState = !TickThread.isTickThreadFor(this.level, this.destroyPos) ? null : this.level.getBlockStateIfLoaded(this.destroyPos); // ShreddedPaper - Don't allow digging into chunks outside our region
             if (blockState == null) {
                 this.isDestroyingBlock = false;
                 return;
diff --git a/net/minecraft/world/level/CustomSpawner.java b/net/minecraft/world/level/CustomSpawner.java
index 7eb6f80df002c9b636aaa6597961524edfd249b2..76f509670783ff580ccff64cd75a7abd3449e6f3 100644
--- a/net/minecraft/world/level/CustomSpawner.java
+++ b/net/minecraft/world/level/CustomSpawner.java
@@ -1,7 +1,7 @@
 package net.minecraft.world.level;
 
 import net.minecraft.server.level.ServerLevel;
-import java.util.function.Supplier
+import java.util.function.Supplier;
 
 public interface CustomSpawner {
     void tick(ServerLevel level, boolean spawnEnemies, boolean spawnFriendlies);
