--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -19,8 +_,11 @@
 import java.util.Optional;
 import java.util.Set;
 import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
 import java.util.function.Function;
 import java.util.function.Predicate;
+
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.ChatFormatting;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.core.BlockPos;
@@ -84,6 +_,7 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.projectile.throwableitemprojectile.ThrownEnderpearl;
 import net.minecraft.world.item.crafting.RecipeManager;
+import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
@@ -491,7 +_,9 @@
         for (ThrownEnderpearl enderpearl : player.getEnderPearls()) {
             // Paper start - Allow using old ender pearl behavior
             if (!enderpearl.level().paperConfig().misc.legacyEnderPearlBehavior) {
+                ShreddedPaper.ensureSync(enderpearl, () -> { // ShreddedPaper - run on ender pearl's thread
                 enderpearl.setRemoved(Entity.RemovalReason.UNLOADED_WITH_PLAYER, org.bukkit.event.entity.EntityRemoveEvent.Cause.PLAYER_QUIT); // CraftBukkit - add Bukkit remove cause
+                }); // ShreddedPaper - run on ender pearl's thread
             }
             // Paper end - Allow using old ender pearl behavior
         }
@@ -516,8 +_,7 @@
         // CraftBukkit start
         // this.broadcastAll(new ClientboundPlayerInfoRemovePacket(List.of(player.getUUID())));
         ClientboundPlayerInfoRemovePacket packet = new ClientboundPlayerInfoRemovePacket(List.of(player.getUUID()));
-        for (int i = 0; i < this.players.size(); i++) {
-            ServerPlayer otherPlayer = this.players.get(i);
+        for (ServerPlayer otherPlayer : this.players) { // ShreddedPaper - Fix concurrent modification
 
             if (otherPlayer.getBukkitEntity().canSee(player.getBukkitEntity())) {
                 otherPlayer.connection.send(packet);
@@ -595,12 +_,13 @@
     }
 
     // Paper start - respawn event
-    public ServerPlayer respawn(final ServerPlayer serverPlayer, final boolean keepAllPlayerData, final Entity.RemovalReason removalReason, final org.bukkit.event.player.PlayerRespawnEvent.RespawnReason respawnReason) {
-        ServerPlayer.RespawnResult result = serverPlayer.findRespawnPositionAndUseSpawnBlock0(!keepAllPlayerData, TeleportTransition.DO_NOTHING, respawnReason);
+    public CompletableFuture<ServerPlayer> respawnAsync(final ServerPlayer serverPlayer, final boolean keepAllPlayerData, final Entity.RemovalReason removalReason, final org.bukkit.event.player.PlayerRespawnEvent.RespawnReason respawnReason) { // ShreddedPaper - respawn async
+        return serverPlayer.findRespawnPositionAndUseSpawnBlock0Async(!keepAllPlayerData, TeleportTransition.DO_NOTHING, respawnReason).thenCompose(result -> { // ShreddedPaper - respawn async
         if (result == null) { // disconnected player during the respawn event
-            return serverPlayer;
+            return CompletableFuture.completedFuture(serverPlayer); // ShreddedPaper - respawn async
         }
         TeleportTransition respawnInfo = result.transition();
+        return CompletableFuture.supplyAsync(() -> { // ShreddedPaper - respawn async
         Level fromLevel = serverPlayer.level();
         // Paper end - respawn event
         this.players.remove(serverPlayer);
@@ -706,6 +_,8 @@
         // Paper end
 
         return player;
+        }, r -> ShreddedPaper.ensureSync(serverPlayer, respawnInfo.newLevel(), ChunkPos.containing(BlockPos.containing(respawnInfo.position())), r)); // ShreddedPaper - respawn async
+        }); // ShreddedPaper - respawn async
     }
 
     public void sendActivePlayerEffects(final ServerPlayer player) {
@@ -970,6 +_,7 @@
         final ResourceKey<Level> dimension,
         final Packet<?> packet
     ) {
+        if (except != null) new io.multipaper.shreddedpaper.event.BroadcastPacketEvent(this.getServer().getLevel(dimension).getWorld(), packet, except.getBukkitEntity()).callEvent(); // ShreddedPaper - BroadcastPacketEvent
         for (int i = 0; i < this.players.size(); i++) {
             ServerPlayer player = this.players.get(i);
             // CraftBukkit start - Test if player receiving packet can see the source of the packet
