From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: caleb <graham.caleb.m@gmail.com>
Date: Fri, 24 Oct 2025 16:53:39 -0500
Subject: [PATCH] update-dependencies


diff --git a/build.gradle.kts b/build.gradle.kts
index ab4e60f07235c499d85402f58cfe0ab47e7410e7..ea05b6b89021f5baa804c4d653700e894267782e 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -73,8 +73,8 @@ dependencies {
     implementation("io.papermc:reflection-rewriter-proxy-generator:$reflectionRewriterVersion")
     // Paper end - Remap reflection
     // Paper start - spark
-    implementation("me.lucko:spark-api:0.1-20240720.200737-2")
-    implementation("me.lucko:spark-paper:1.10.99-SNAPSHOT")
+    implementation("me.lucko:spark-api:0.1-SNAPSHOT")
+    implementation("me.lucko:spark-paper:1.10.133-SNAPSHOT")
     // Paper end - spark
 }
 
