From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: caleb <graham.caleb.m@gmail.com>
Date: Sun, 26 Oct 2025 01:15:34 -0500
Subject: [PATCH] Update Imports and Misc Fixes


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
index 3305fcb8d2bfbd0dfff9f1fe64f508907431404f..e0ad061e0bcd384cedb202f53e5a5d88efaf9641 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
@@ -9,6 +9,7 @@ import java.util.OptionalInt;
 import java.util.jar.Manifest;
 import net.kyori.adventure.key.Key;
 import net.minecraft.SharedConstants;
+import net.kyori.adventure.key.Key;
 import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.craftbukkit.Main;
 import org.jetbrains.annotations.NotNull;
@@ -34,6 +35,8 @@ public record ServerBuildInfoImpl(
     private static final String BRAND_PUFFERFISH_NAME = "Pufferfish"; // Purpur
     private static final String BRAND_PURPUR_NAME = "Purpur"; // Purpur
     private static final String BRAND_SHREDDEDPAPER_NAME = "ShreddedPaper"; // Purpur // ShreddedPaper
+    private static final Key BRAND_SHREDDEDPAPER_ID = Key.key("shreddedpaper");
+
 
     private static final String BUILD_DEV = "DEV";
 
@@ -46,7 +49,7 @@ public record ServerBuildInfoImpl(
             getManifestAttribute(manifest, ATTRIBUTE_BRAND_ID)
                 .map(Key::key)
                 .orElse(BRAND_SHREDDEDPAPER_ID), // Purpur // ShreddedPaper
-            getManifestAttribute(manifest, ATTRIBUTE_BRAND_NAME)
+        getManifestAttribute(manifest, ATTRIBUTE_BRAND_NAME)
                 .orElse(BRAND_SHREDDEDPAPER_NAME), // Purpur // ShreddedPaper
             SharedConstants.getCurrentVersion().getId(),
             SharedConstants.getCurrentVersion().getName(),
diff --git a/src/main/java/net/minecraft/server/level/ServerChunkCache.java b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
index 1886b11a63613da237c3c780d7a46ed44892359c..d2d644235ef08c716f4d477e3530ca6e7d6cb6e3 100644
--- a/src/main/java/net/minecraft/server/level/ServerChunkCache.java
+++ b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
@@ -507,7 +507,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
                 //gameprofilerfiller.popPush("spawnAndTick"); // Purpur
                 boolean flag = this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING) && !this.level.players().isEmpty(); // CraftBukkit
 
-                if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) Util.shuffle(list, this.level.random); // Paper - per player mob spawns - do not need this when per-player is enabled
+                if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) Util.shuffle(this.level.players(), this.level.random); // Paper - per player mob spawns - do not need this when per-player is enabled
                 // Paper start - PlayerNaturallySpawnCreaturesEvent
                 int chunkRange = level.spigotConfig.mobSpawnRange;
                 chunkRange = (chunkRange > level.spigotConfig.viewDistance) ? (byte) level.spigotConfig.viewDistance : chunkRange;
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 1cccc9ce0c0ecf90559f0bad820ac05d7f13205d..a02540d121a396897d28c8cce503d2e311f960ea 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -80,6 +80,7 @@ import net.minecraft.util.Mth;
 import net.minecraft.util.ProgressListener;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.Unit;
+import net.minecraft.util.datafix.DataFixTypes;
 import net.minecraft.util.valueproviders.IntProvider;
 import net.minecraft.util.valueproviders.UniformInt;
 import net.minecraft.world.DifficultyInstance;
@@ -133,9 +134,11 @@ import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.chunk.LevelChunkSection;
 import net.minecraft.world.level.chunk.storage.EntityStorage;
 import net.minecraft.world.level.chunk.storage.RegionStorageInfo;
+import net.minecraft.world.level.chunk.storage.SimpleRegionStorage;
 import net.minecraft.world.level.dimension.BuiltinDimensionTypes;
 import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.dimension.end.EndDragonFight;
+import net.minecraft.world.level.entity.EntityPersistentStorage;
 import net.minecraft.world.level.entity.EntityTypeTest;
 import net.minecraft.world.level.entity.LevelCallback;
 import net.minecraft.world.level.entity.LevelEntityGetter;
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index cbf09786d70d1fe2b0fcbbb878961fcd175b6a10..6fe877b95311cd00c46d52614edfa84bd7a8a09e 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1,5 +1,6 @@
 package net.minecraft.server.level;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import com.google.common.net.InetAddresses;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.util.Either;
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 5d79cc7a64e11ac113868e7ddb2dba04721e44a3..a73df22ca0a021383df3734aa338adb5bda2084b 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -8,6 +8,7 @@ import com.google.common.collect.Sets;
 import com.google.common.collect.UnmodifiableIterator;
 import com.mojang.logging.LogUtils;
 import io.multipaper.shreddedpaper.ShreddedPaper;
+import io.multipaper.shreddedpaper.region.RegionPos;
 import it.unimi.dsi.fastutil.doubles.DoubleList;
 import it.unimi.dsi.fastutil.doubles.DoubleListIterator;
 import it.unimi.dsi.fastutil.floats.FloatArraySet;
