From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: caleb <graham.caleb.m@gmail.com>
Date: Tue, 28 Oct 2025 13:37:55 -0500
Subject: [PATCH] Port patches through Optimize "tracker-full-update-frequency
 further"


diff --git a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
index b1b4415fe64a36778beb31938ae5d360bc936d02..0757aeb2c3aa48dcf1e02ffeebbad43f8c4e571c 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
@@ -75,6 +75,10 @@ public class TickThread extends Thread {
     }
 
     public static void ensureTickThread(final Level world, final BlockPos pos, final String reason) {
+        if (Thread.currentThread() == world.getServer().getRunningThread()) {
+            return; // Allow main server thread
+        }
+
         if (!isTickThreadFor(world, pos)) {
             failedTickThreadCheck(reason, "world=" + ((ServerLevel) world).convertable.getLevelId() + ", pos=" + pos);
         }
diff --git a/src/main/java/io/multipaper/shreddedpaper/config/ShreddedPaperConfiguration.java b/src/main/java/io/multipaper/shreddedpaper/config/ShreddedPaperConfiguration.java
index 99ff58475641bc12fa3a3fd99afd6f15ac315957..9d03b9607cebb746cd01780a15640529d008ea73 100644
--- a/src/main/java/io/multipaper/shreddedpaper/config/ShreddedPaperConfiguration.java
+++ b/src/main/java/io/multipaper/shreddedpaper/config/ShreddedPaperConfiguration.java
@@ -41,8 +41,24 @@ public class ShreddedPaperConfiguration extends ConfigurationPart {
     public class Optimizations extends ConfigurationPart {
 
         public int entityActivationCheckFrequency = 20;
+        public boolean disableVanishApi = false;
+        public boolean useLazyExecuteWhenNotFlushing = true;
+        public boolean processTrackQueueInParallel = true;
+        public boolean flushQueueInParallel = true;
+        public int maximumTrackersPerEntity = 500;
+        public long trackerFullUpdateFrequency = 20;
+        public boolean writePlayerSavesAsync = true;
+        public ChunkPacketCaching chunkPacketCaching;
+
+        public class ChunkPacketCaching extends ConfigurationPart {
+
+            public boolean enabled = true;
+            public boolean useSoftReferences = true;
+            public long expireAfter = 1200;
 
-    }
 
+        }
+
+    }
 
 }
diff --git a/src/main/java/io/multipaper/shreddedpaper/permissions/ShreddedPaperPermissions.java b/src/main/java/io/multipaper/shreddedpaper/permissions/ShreddedPaperPermissions.java
index 4999f297b04d151e8000fdc309081b5cee8b11ae..a72566749fce7c3e0a89e16afc31caf01117034a 100644
--- a/src/main/java/io/multipaper/shreddedpaper/permissions/ShreddedPaperPermissions.java
+++ b/src/main/java/io/multipaper/shreddedpaper/permissions/ShreddedPaperPermissions.java
@@ -1,6 +1,7 @@
 package io.multipaper.shreddedpaper.permissions;
 
 import org.bukkit.permissions.Permission;
+import org.bukkit.permissions.PermissionDefault;
 import org.bukkit.util.permissions.DefaultPermissions;
 
 public class ShreddedPaperPermissions {
@@ -9,6 +10,8 @@ public class ShreddedPaperPermissions {
     public static void registerCorePermissions() {
         Permission parent = DefaultPermissions.registerPermission(ROOT, "Gives the user the ability to use all ShreddedPaper utilities and commands");
 
+        DefaultPermissions.registerPermission(ROOT + ".maximumtrackerbypass", "Gives the user the ability to see all entities regardless of the maximum tracker config", PermissionDefault.OP, parent);
+
         ShreddedPaperCommandPermissions.registerPermissions(parent);
 
         parent.recalculatePermissibles();
diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
index 57c8aeb70bf394f17faf23350d913d27ec513130..2815b3333174263da9cb69fe0eea98f24adc205c 100644
--- a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
@@ -1,9 +1,13 @@
 package io.multipaper.shreddedpaper.threading;
 
+import com.google.common.collect.Lists;
 import com.mojang.logging.LogUtils;
+import io.multipaper.shreddedpaper.config.ShreddedPaperConfiguration;
 import io.multipaper.shreddedpaper.region.RegionPos;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
@@ -24,33 +28,72 @@ public class ShreddedPaperChunkTicker {
 
     public static boolean tickingChunks = false;
 
+    private static final List<Entity> trackedEntitiesWorkerList = new ArrayList<>(); // Re-usable list for processing tracked entities in parallel
+
     public static void tickChunks(ServerLevel level, NaturalSpawner.SpawnState spawnState) {
         final List<CompletableFuture<Void>> futures = new ArrayList<>();
+
         tickingChunks = true;
 
-        level.chunkSource.tickingRegions.forEach(region -> {
-            LOGGER.debug("[ShreddedTicker] START region {}", region.getRegionPos());
+        level.chunkSource.tickingRegions.forEach(
+                region -> futures.add(tickRegion(level, region, spawnState))
+        );
 
-            CompletableFuture<Void> f = tickRegion(level, region, spawnState)
-                    .orTimeout(30, TimeUnit.SECONDS)
-                    .whenComplete((v, ex) -> {
-                        if (ex != null) {
-                            LOGGER.error("[ShreddedTicker] FAIL  region {}", region.getRegionPos(), ex);
-                        } else {
-                            LOGGER.debug("[ShreddedTicker] DONE  region {}", region.getRegionPos());
-                        }
-                    });
+        waitForFutures(level, futures);
 
-            futures.add(f);
-        });
+        if (ShreddedPaperConfiguration.get().optimizations.processTrackQueueInParallel) processTrackQueueInParallel(level);
 
-        // wait once on all regions
-        CompletableFuture<Void> all = CompletableFuture.allOf(futures.toArray(CompletableFuture[]::new));
-        if (!all.isDone()) {
-            level.getChunkSource().mainThreadProcessor.managedBlock(all::isDone);
-        }
+
+        if (ShreddedPaperConfiguration.get().optimizations.flushQueueInParallel) flushQueueInParallel(level);
 
         tickingChunks = false;
+
+        MinecraftServer.getServer().executeMidTickTasks();
+    }
+
+    private static void waitForFutures(ServerLevel level, Iterable<CompletableFuture<Void>> futures) {
+        for (CompletableFuture<Void> future : futures) {
+            if (!future.isDone()) {
+                level.getChunkSource().mainThreadProcessor.managedBlock(future::isDone);
+            }
+            if (future.isCompletedExceptionally()) {
+                throw new RuntimeException("Exception while ticking world %s".formatted(level.getWorld().getName()), future.exceptionNow());
+            }
+        }
+    }
+
+    private static void processTrackQueueInParallel(ServerLevel level) {
+        level.getChunkSource().mainThreadProcessor.managedBlock(() -> level.chunkScheduler.getRegionLocker().globalLock().tryWriteLock() != 0);
+        try {
+            List<CompletableFuture<Void>> futures = new ArrayList<>();
+
+            trackedEntitiesWorkerList.clear();
+            level.chunkSource.tickingRegions.forEach(
+                    region -> region.forEachTrackedEntity(trackedEntitiesWorkerList::add)
+            );
+
+            List<List<Entity>> trackedEntitiesTasks = Lists.partition(trackedEntitiesWorkerList, Math.max(1, trackedEntitiesWorkerList.size() / ShreddedPaperTickThread.THREAD_COUNT / 3));
+            for (List<Entity> trackedEntities : trackedEntitiesTasks) {
+                if (trackedEntities.isEmpty()) continue;
+                futures.add(CompletableFuture.runAsync(() -> trackedEntities.forEach(ShreddedPaperEntityTicker::processTrackQueue), ShreddedPaperTickThread.getExecutor()));
+            }
+
+            waitForFutures(level, futures);
+        } finally {
+            level.chunkScheduler.getRegionLocker().globalLock().tryUnlockWrite();
+        }
+    }
+
+    private static void flushQueueInParallel(ServerLevel level) {
+        List<CompletableFuture<Void>> futures = new ArrayList<>();
+
+        List<List<ServerPlayer>> playersTasks = Lists.partition(new ArrayList<>(level.players()), Math.max(1, level.players().size() / ShreddedPaperTickThread.THREAD_COUNT / 3));
+        for (List<ServerPlayer> players : playersTasks) {
+            if (players.isEmpty()) continue;
+            futures.add(CompletableFuture.runAsync(() -> players.forEach(player -> player.connection.connection.flushQueue()), ShreddedPaperTickThread.getExecutor()));
+        }
+
+        waitForFutures(level, futures);
     }
 
     public static CompletableFuture<Void> tickRegion(ServerLevel level, LevelChunkRegion region, NaturalSpawner.SpawnState spawnercreature_d) {
@@ -130,6 +173,12 @@ public class ShreddedPaperChunkTicker {
     private static void _tickChunk(ServerLevel level, LevelChunk chunk1, NaturalSpawner.SpawnState spawnercreature_d) {
         if (chunk1.moonrise$getChunkAndHolder().holder().hasChangedSections) ShreddedPaperChunkChangesBroadcaster.setNeedsBroadcast(chunk1); // ShreddedPaper - ensure queued for broadcasting changes
 
+        // ShreddedPaper start - clear chunk packet cache
+        if (chunk1.cachedChunkPacket != null && chunk1.cachedChunkPacketLastAccessed < level.getGameTime() - ShreddedPaperConfiguration.get().optimizations.chunkPacketCaching.expireAfter) {
+            chunk1.cachedChunkPacket = null;
+        }
+        // ShreddedPaper end - clear chunk packet cache
+
         // Start - Import the same variables as the original chunk ticking method to make copying new changes easier
         int j = 1; // Inhabited time increment in ticks
         boolean flag = level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING) && !level.players().isEmpty(); // Should run mob spawning code
diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperRegionLocker.java b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperRegionLocker.java
index ef32faef3bc833cf0b461f50255d316300d06f56..dcaa8e7dbe33994dff85c57571a5d73f9cd72dfd 100644
--- a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperRegionLocker.java
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperRegionLocker.java
@@ -5,12 +5,10 @@ import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
 import io.multipaper.shreddedpaper.region.RegionPos;
 import io.multipaper.shreddedpaper.util.ObjectHolder;
 
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.locks.LockSupport;
+import java.util.concurrent.locks.StampedLock;
 import java.util.function.Supplier;
 
 /**
@@ -28,6 +26,7 @@ public class ShreddedPaperRegionLocker {
     public static final int REGION_LOCK_RADIUS = 1;
 
     private final ConcurrentHashMap<RegionPos, Thread> lockedRegions = new ConcurrentHashMap<>();
+    private final StampedLock globalLock = new StampedLock();
 
     private final ThreadLocal<Set<RegionPos>> localLocks = ThreadLocal.withInitial(ObjectOpenHashSet::new);
     private final ThreadLocal<Set<RegionPos>> readOnlyLocks = ThreadLocal.withInitial(ObjectOpenHashSet::new);
@@ -166,4 +165,13 @@ public class ShreddedPaperRegionLocker {
         }
     }
 
+
+    /**
+     * globalLock.writeLock() will claim all regions
+     */
+    public StampedLock globalLock() {
+        return this.globalLock;
+    }
+
+
 }
diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperTickThread.java b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperTickThread.java
index 15561441047c77f2ed6d4c69134965fefcb1c290..2be127bf10244c70804941f3cfefd200c175e4e4 100644
--- a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperTickThread.java
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperTickThread.java
@@ -13,6 +13,8 @@ public class ShreddedPaperTickThread extends TickThread {
 
     private static final Logger LOGGER = LogUtils.getClassLogger();
 
+    public static final int THREAD_COUNT;
+
     private static final AtomicInteger ID_GENERATOR = new AtomicInteger();
 
     private static final ScheduledThreadPoolExecutor executor;
@@ -24,9 +26,9 @@ public class ShreddedPaperTickThread extends TickThread {
             threadCount = Math.max(Runtime.getRuntime().availableProcessors() - 1, 1);
         }
 
-        executor = new ScheduledThreadPoolExecutor(threadCount, r -> new ShreddedPaperTickThread(r, "ShreddedPaperTickThread-%d"));
-
-        LOGGER.info("Using {} threads", threadCount);
+        THREAD_COUNT = threadCount;
+        executor = new ScheduledThreadPoolExecutor(THREAD_COUNT, r -> new ShreddedPaperTickThread(r, "ShreddedPaperTickThread-%d"));
+        LOGGER.info("Using {} threads", THREAD_COUNT);
     }
 
     public ShreddedPaperTickThread(Runnable run, String name) {
diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java b/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java
index 91845c88b5ff01dadd29a5131c714dad93ead153..3483e6bfcb7b7d076a1d5767efa9127b75ed86dd 100644
--- a/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java
@@ -42,7 +42,9 @@ public class SynchronousPluginExecution {
     }
 
     public static void execute(Plugin plugin, RunnableWithException runnable) throws Exception {
-        if (plugin == null || !ShreddedPaperConfiguration.get().multithreading.runUnsupportedPluginsInSync || plugin.getDescription().isFoliaSupported() || TickThread.isShutdownThread()) {
+        //port patch from UnlimitedBytes
+        ShreddedPaperConfiguration config = ShreddedPaperConfiguration.get();
+        if (plugin == null || config == null || !config.multithreading.runUnsupportedPluginsInSync || plugin.getDescription().isFoliaSupported() || TickThread.isShutdownThread()) {
             // Multi-thread safe plugin, run it straight away
             runnable.run();
             return;
diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
index bbdaaae914259be74c0e5f4ce42272a3858b8283..e6b1ff4f51afd9c88d766ba0594c8739019eb1d9 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
@@ -42,7 +42,7 @@ class PaperEventManager {
     public void callEvent(@NotNull Event event) {
         if (event.isAsynchronous() && (Thread.currentThread() instanceof TickThread || ShreddedPaperTickThread.isShreddedPaperTickThread())) {
             throw new IllegalStateException(event.getEventName() + " may only be triggered asynchronously.");
-        } else if (!event.isAsynchronous() && !this.server.isPrimaryThread() && !this.server.isStopping()) {
+        } else if (!event.isAsynchronous() && this.server != null && !this.server.isPrimaryThread() && !this.server.isStopping()) {
             throw new IllegalStateException(event.getEventName() + " may only be triggered synchronously.");
         }
 
diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index e99caffaac884c0bbbc03ad2aff119c7c737f632..ad60f7a77d2659798c332e26ca272e5120cccbce 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -5,6 +5,7 @@ import com.google.common.base.Suppliers;
 import com.google.common.collect.Queues;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import com.mojang.logging.LogUtils;
+import io.multipaper.shreddedpaper.config.ShreddedPaperConfiguration;
 import io.netty.bootstrap.Bootstrap;
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelException;
@@ -33,6 +34,7 @@ import io.netty.channel.socket.nio.NioSocketChannel;
 import io.netty.handler.flow.FlowControlHandler;
 import io.netty.handler.timeout.ReadTimeoutHandler;
 import io.netty.handler.timeout.TimeoutException;
+import io.netty.util.concurrent.AbstractEventExecutor;
 import java.net.InetSocketAddress;
 import java.net.SocketAddress;
 import java.nio.channels.ClosedChannelException;
@@ -451,7 +453,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
     public void send(Packet<?> packet, @Nullable PacketSendListener callbacks, boolean flush) {
         // Paper start - Optimize network: Handle oversized packets better
-        final boolean connected = this.isConnected();
+        final boolean connected = true || this.isConnected(); // ShreddedPaper - Checking if the player is connected is expensive
         if (!connected && !this.preparing) {
             return;
         }
@@ -500,6 +502,14 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
         if (this.channel.eventLoop().inEventLoop()) {
             this.doSendPacket(packet, callbacks, flush);
         } else {
+            // ShreddedPaper start - Use lazyExecute if we aren't flushing
+            if (ShreddedPaperConfiguration.get().optimizations.useLazyExecuteWhenNotFlushing && !flush) {
+                ((AbstractEventExecutor) this.channel.eventLoop()).lazyExecute(() -> {
+                    this.doSendPacket(packet, callbacks, flush);
+                });
+                return;
+            }
+            // ShreddedPaper end - Use lazyExecute if we aren't flushing
             this.channel.eventLoop().execute(() -> {
                 this.doSendPacket(packet, callbacks, flush);
             });
@@ -569,7 +579,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     }
 
     // Paper start - Optimize network: Rewrite this to be safer if ran off main thread
-    private boolean flushQueue() {
+    public boolean flushQueue() {
         if (!this.isConnected()) {
             return true;
         }
diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
index cfcac0fdc130120cb1f8d97c6353d93db7ddf81b..0b75fe53817f291b80a80f5fe4432bca256d65bb 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
@@ -20,6 +20,7 @@ public class ClientboundLevelChunkWithLightPacket implements Packet<ClientGamePa
     private final ClientboundLightUpdatePacketData lightData;
     // Paper start - Async-Anti-Xray - Ready flag for the connection
     private volatile boolean ready;
+    public boolean modifyBlocks; // ShreddedPaper
 
     @Override
     public boolean isReady() {
@@ -37,6 +38,7 @@ public class ClientboundLevelChunkWithLightPacket implements Packet<ClientGamePa
         ChunkPos chunkPos = chunk.getPos();
         this.x = chunkPos.x;
         this.z = chunkPos.z;
+        this.modifyBlocks = modifyBlocks; // ShreddedPaper
         com.destroystokyo.paper.antixray.ChunkPacketInfo<net.minecraft.world.level.block.state.BlockState> chunkPacketInfo = modifyBlocks ? chunk.getLevel().chunkPacketBlockController.getChunkPacketInfo(this, chunk) : null;
         this.chunkData = new ClientboundLevelChunkPacketData(chunk, chunkPacketInfo);
         // Paper end
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 1717f6e9b2d46b2867aec0bb51f79e873db9672e..8d1b5f546842833684dd234784c68d34baa433eb 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1570,6 +1570,48 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         return flag;
     }
 
+    public final void executeMidTickTasks() {
+        org.spigotmc.AsyncCatcher.catchOp("mid tick chunk task execution");
+        long startTime = System.nanoTime();
+        if ((startTime - lastMidTickExecute) <= CHUNK_TASK_QUEUE_BACKOFF_MIN_TIME || (startTime - lastMidTickExecuteFailure) <= TASK_EXECUTION_FAILURE_BACKOFF) {
+            // it's shown to be bad to constantly hit the queue (chunk loads slow to a crawl), even if no tasks are executed.
+            // so, backoff to prevent this
+            return;
+        }
+
+        // co.aikar.timings.MinecraftTimings.midTickChunkTasks.startTiming(); // Purpur
+        executingMidTickTasks.set(true); // ShreddedPaper
+        try {
+            for (;;) {
+                boolean moreTasks = this.tickMidTickTasks();
+                long currTime = System.nanoTime();
+                long diff = currTime - startTime;
+
+                if (!moreTasks || diff >= MAX_CHUNK_EXEC_TIME) {
+                    if (!moreTasks) {
+                        lastMidTickExecuteFailure = currTime;
+                    }
+
+                    // note: negative values reduce the time
+                    long overuse = diff - MAX_CHUNK_EXEC_TIME;
+                    if (overuse >= (10L * 1000L * 1000L)) { // 10ms
+                        // make sure something like a GC or dumb plugin doesn't screw us over...
+                        overuse = 10L * 1000L * 1000L; // 10ms
+                    }
+
+                    double overuseCount = (double)overuse/(double)MAX_CHUNK_EXEC_TIME;
+                    long extraSleep = (long)Math.round(overuseCount*CHUNK_TASK_QUEUE_BACKOFF_MIN_TIME);
+
+                    lastMidTickExecute = currTime + extraSleep;
+                    return;
+                }
+            }
+        } finally {
+            executingMidTickTasks.set(false); // ShreddedPaper
+            // co.aikar.timings.MinecraftTimings.midTickChunkTasks.stopTiming(); // Purpur
+        }
+    }
+
     private boolean pollTaskInternal() {
         if (super.pollTask()) {
             this.moonrise$executeMidTickTasks(); // Paper - rewrite chunk system
diff --git a/src/main/java/net/minecraft/server/level/ChunkHolder.java b/src/main/java/net/minecraft/server/level/ChunkHolder.java
index 7c9cbaa3983c3bb48fba7d033f1d1ed8c99b25da..a8eb8f353155b44b3fdb96fd656ee16665e2b996 100644
--- a/src/main/java/net/minecraft/server/level/ChunkHolder.java
+++ b/src/main/java/net/minecraft/server/level/ChunkHolder.java
@@ -282,6 +282,7 @@ public class ChunkHolder extends GenerationChunkHolder implements ca.spottedleaf
 
     public void broadcastChanges(LevelChunk chunk) {
         if (this.hasChangedSections || !this.skyChangedLightSectionFilter.isEmpty() || !this.blockChangedLightSectionFilter.isEmpty()) {
+            if (chunk.cachedChunkPacket != null) chunk.cachedChunkPacket = null; // ShreddedPaper
             Level world = chunk.getLevel();
             List list;
 
diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
index a8c126c2b9d55445f8f080774335d171eef66c79..2c489ac70a95b2cb934b385b5f757b09cf281e6d 100644
--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
+++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
@@ -10,6 +10,8 @@ import com.google.common.collect.Queues;
 import com.google.common.collect.Sets;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
+import io.multipaper.shreddedpaper.config.ShreddedPaperConfiguration;
+import io.multipaper.shreddedpaper.threading.ShreddedPaperTickThread;
 import io.multipaper.shreddedpaper.util.Int2ObjectMapWrapper;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
 import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
@@ -22,12 +24,16 @@ import it.unimi.dsi.fastutil.longs.Long2ObjectMap.Entry;
 import it.unimi.dsi.fastutil.longs.LongIterator;
 import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
 import it.unimi.dsi.fastutil.longs.LongSet;
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import it.unimi.dsi.fastutil.objects.ObjectBidirectionalIterator;
 import it.unimi.dsi.fastutil.objects.ObjectIterator;
+import it.unimi.dsi.fastutil.objects.ObjectLinkedOpenHashSet;
 import java.io.IOException;
 import java.io.Writer;
 import java.nio.file.Path;
 import java.util.ArrayList;
+import java.util.Collection;
+import java.util.ConcurrentModificationException;
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
@@ -37,18 +43,21 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.Queue;
 import java.util.Set;
+import java.util.TreeSet;
 import java.util.concurrent.CancellationException;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionException;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReentrantLock;
 import java.util.function.BooleanSupplier;
 import java.util.function.Consumer;
 import java.util.function.IntFunction;
 import java.util.function.IntSupplier;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
+import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportedException;
@@ -104,6 +113,7 @@ import net.minecraft.world.level.storage.DimensionDataStorage;
 import net.minecraft.world.level.storage.LevelStorageSource;
 import net.minecraft.world.phys.Vec3;
 import org.apache.commons.lang3.mutable.MutableBoolean;
+import org.jetbrains.annotations.NotNull;
 import org.slf4j.Logger;
 
 // CraftBukkit start
@@ -889,6 +899,7 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
                         ServerPlayer entityplayer = (ServerPlayer) entity;
 
                         this.updatePlayerStatus(entityplayer, true);
+                        if (true) return; // ShreddedPaper - handled elsewhere
                         Iterator objectiterator = this.entityConcurrentMap.values().iterator(); // ShreddedPaper - use entityConcurrentMap
 
                         while (objectiterator.hasNext()) {
@@ -1115,6 +1126,7 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
         final Entity entity;
         private final int range;
         SectionPos lastSectionPos;
+        private final ReentrantLock seenByLock = new ReentrantLock(); // ShreededPaper - seenBy thread safety
         public final Set<ServerPlayerConnection> seenBy = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>(); // Paper - Perf: optimise map impl
 
         // Paper start - optimise entity tracker
@@ -1202,6 +1214,49 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
             this.lastSectionPos = SectionPos.of((EntityAccess) entity);
         }
 
+        ObjectLinkedOpenHashSet<ServerPlayer> lastTrackerCandidates;
+
+        public final void updatePlayers(ObjectLinkedOpenHashSet<ServerPlayer> newTrackerCandidates) { // ShreddedPaper - make public
+            ObjectLinkedOpenHashSet<ServerPlayer> oldTrackerCandidates = this.lastTrackerCandidates;
+            this.lastTrackerCandidates = newTrackerCandidates;
+
+            boolean fullTrackerUpdate = (level.getGameTime() ^ (long) this.entity.getId()) % ShreddedPaperConfiguration.get().optimizations.trackerFullUpdateFrequency == 0L; // ShreddedPaper - trackerFullUpdateFrequency
+            if (fullTrackerUpdate && this.entity instanceof ServerPlayer player) player.hasMaximumTrackerBypassPermission = player.getBukkitEntity().hasPermission("shreddedpaper.maximumtrackerbypass"); // ShreddedPaper
+            if (this.lastTrackerCandidates != null && this.lastTrackerCandidates.size() > ShreddedPaperConfiguration.get().optimizations.maximumTrackersPerEntity) { updatePlayersLimitedAndOrdered(oldTrackerCandidates != newTrackerCandidates && (oldTrackerCandidates == null || newTrackerCandidates == null || oldTrackerCandidates.size() != newTrackerCandidates.size()), fullTrackerUpdate); return; } // ShreddedPaper - maximum trackers per entity
+
+            if (newTrackerCandidates != null) {
+                Object[] rawData = newTrackerCandidates.toArray(new ServerPlayer[newTrackerCandidates.size()]);
+                for (int i = 0, len = rawData.length; i < len; ++i) {
+                    Object raw = rawData[i];
+                    if (!(raw instanceof ServerPlayer)) {
+                        continue;
+                    }
+                    ServerPlayer player = (ServerPlayer)raw;
+                    if (fullTrackerUpdate || !this.seenBy.contains(player.connection)) this.updatePlayer(player); // ShreddedPaper - trackerFullUpdateFrequency
+                }
+            }
+
+            if (!fullTrackerUpdate) return; // ShreddedPaper - trackerFullUpdateFrequency
+
+            if (oldTrackerCandidates == newTrackerCandidates) {
+                // this is likely the case.
+                // means there has been no range changes, so we can just use the above for tracking.
+                return;
+            }
+
+            // stuff could have been removed, so we need to check the trackedPlayers set
+            // for players that were removed
+
+            seenByLock.lock(); try { // ShreededPaper - seenBy thread safety
+                for (ServerPlayerConnection conn : this.seenBy.toArray(new ServerPlayerConnection[0])) { // avoid CME
+                    if (newTrackerCandidates == null || !newTrackerCandidates.contains(conn.getPlayer())) {
+                        this.updatePlayer(conn.getPlayer());
+                    }
+                }
+            } finally { seenByLock.unlock(); } // ShreededPaper - seenBy thread safety
+        }
+        // Paper end - use distance map to optimise tracker
+
         public boolean equals(Object object) {
             return object instanceof ChunkMap.TrackedEntity ? ((ChunkMap.TrackedEntity) object).entity.getId() == this.entity.getId() : false;
         }
@@ -1292,6 +1347,94 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
             }
         }
 
+        // ShreddedPaper start - maximum trackers per entity
+        private record DistancedServerPlayer(ServerPlayer serverPlayer, double distanceSqr, boolean hasBypassPerm) implements Comparable<DistancedServerPlayer> {
+            DistancedServerPlayer(ServerPlayer serverPlayer, Entity entity) {
+                this(serverPlayer, distanceBetween(serverPlayer, entity), serverPlayer.hasMaximumTrackerBypassPermission);
+            }
+
+            private static double distanceBetween(ServerPlayer serverPlayer, Entity entity) {
+                double vec3d_dx = serverPlayer.getX() - entity.getX();
+                double vec3d_dz = serverPlayer.getZ() - entity.getZ();
+                return vec3d_dx * vec3d_dx + vec3d_dz * vec3d_dz;
+            }
+
+            @Override
+            public int compareTo(@NotNull ChunkMap.TrackedEntity.DistancedServerPlayer o) {
+                int compareResult = Double.compare(hasBypassPerm ? 0 : this.distanceSqr, o.hasBypassPerm ? 0 : o.distanceSqr);
+                return compareResult == 0 ? Integer.compare(this.serverPlayer.getId(), o.serverPlayer.getId()) : compareResult;
+            }
+        }
+
+        public void updatePlayersLimitedAndOrdered(boolean hasTrackerCandidatesChanged, boolean fullTrackerUpdate) {
+            if (!hasTrackerCandidatesChanged && !fullTrackerUpdate) return;
+
+            TreeSet<DistancedServerPlayer> playerSet = new TreeSet<>();
+            Object[] rawData = this.lastTrackerCandidates.toArray(new ServerPlayer[lastTrackerCandidates.size()]);
+
+            for (int index = 0, len = rawData.length; index < len; ++index) {
+                Object raw = rawData[index];
+                if (!(raw instanceof ServerPlayer player)) {
+                    continue;
+                }
+                if (player == this.entity) continue;
+                DistancedServerPlayer distancedPlayer = new DistancedServerPlayer(player, this.entity);
+                int i = ChunkMap.this.getPlayerViewDistance(player);
+                double d0 = (double) Math.min(this.getEffectiveRange(), i * 16);
+                double d2 = d0 * d0;
+                boolean flag = distancedPlayer.distanceSqr <= d2 && this.entity.broadcastToPlayer(player) && ChunkMap.this.isChunkTracked(player, this.entity.chunkPosition().x, this.entity.chunkPosition().z);
+                // Paper start - Configurable entity tracking range by Y
+                if (flag && level.paperConfig().entities.trackingRangeY.enabled) {
+                    double rangeY = level.paperConfig().entities.trackingRangeY.get(this.entity, -1);
+                    if (rangeY != -1) {
+                        double vec3d_dy = player.getY() - this.entity.getY();
+                        flag = vec3d_dy * vec3d_dy <= rangeY * rangeY;
+                    }
+                }
+                // Paper end - Configurable entity tracking range by Y
+
+                // CraftBukkit start - respect vanish API
+                if (flag && !player.getBukkitEntity().canSee(this.entity.getBukkitEntity())) { // Paper - only consider hits
+                    flag = false;
+                }
+
+                if (flag) {
+                    playerSet.add(distancedPlayer);
+                }
+            }
+
+            seenByLock.lock();
+            try {
+                Set<ServerPlayer> canBeSeenBy = new ReferenceOpenHashSet<>(Math.min(ShreddedPaperConfiguration.get().optimizations.maximumTrackersPerEntity, playerSet.size()));
+                Iterator<DistancedServerPlayer> playerSetIterator = playerSet.iterator();
+                int count = 0;
+                while (playerSetIterator.hasNext() && count++ < ShreddedPaperConfiguration.get().optimizations.maximumTrackersPerEntity) { // ShreddedPaper - maximum trackers per entity
+                    DistancedServerPlayer distancedPlayer = playerSetIterator.next();
+                    ServerPlayer player = distancedPlayer.serverPlayer();
+                    canBeSeenBy.add(player);
+                    if (this.seenBy.add(player.connection)) {
+                        // Paper start - entity tracking events
+                        if (io.papermc.paper.event.player.PlayerTrackEntityEvent.getHandlerList().getRegisteredListeners().length == 0 || new io.papermc.paper.event.player.PlayerTrackEntityEvent(player.getBukkitEntity(), this.entity.getBukkitEntity()).callEvent()) {
+                            this.serverEntity.addPairing(player);
+                        }
+                        // Paper end - entity tracking events
+                        this.serverEntity.onPlayerAdd(); // Paper - fix desync when a player is added to the tracker
+                    }
+                }
+
+                if (this.seenBy.size() != canBeSeenBy.size()) {
+                    for (ServerPlayerConnection conn : this.seenBy.toArray(new ServerPlayerConnection[0])) { // avoid CME
+                        if (!canBeSeenBy.contains(conn.getPlayer()) && this.seenBy.remove(conn)) {
+                            this.serverEntity.removePairing(conn.getPlayer());
+                        }
+                    }
+                }
+            } finally {
+                seenByLock.unlock();
+            }
+        }
+        // ShreddedPaper end - maximum trackers per entity
+
         private int scaledRange(int initialDistance) {
             return ChunkMap.this.level.getServer().getScaledTrackingDistance(initialDistance);
         }
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 6fe877b95311cd00c46d52614edfa84bd7a8a09e..2f3d7dc1f05a8cb54f4878bd1076d32441bcdc68 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -261,6 +261,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
     private BlockPos respawnPosition;
     private boolean respawnForced;
     private float respawnAngle;
+    public boolean hasMaximumTrackerBypassPermission; // ShreddedPaper
     private final TextFilter textFilter;
     private boolean textFilteringEnabled;
     private boolean allowsListing;
diff --git a/src/main/java/net/minecraft/server/network/PlayerChunkSender.java b/src/main/java/net/minecraft/server/network/PlayerChunkSender.java
index dafa2cf7d3c49fc5bdcd68d2a952812774a1dfe4..8eee7b8c5c8315d51186925924d74df8774524c2 100644
--- a/src/main/java/net/minecraft/server/network/PlayerChunkSender.java
+++ b/src/main/java/net/minecraft/server/network/PlayerChunkSender.java
@@ -2,8 +2,13 @@ package net.minecraft.server.network;
 
 import com.google.common.collect.Comparators;
 import com.mojang.logging.LogUtils;
+import io.multipaper.shreddedpaper.config.ShreddedPaperConfiguration;
 import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
 import it.unimi.dsi.fastutil.longs.LongSet;
+
+import java.lang.ref.Reference;
+import java.lang.ref.SoftReference;
+import java.lang.ref.WeakReference;
 import java.util.Comparator;
 import java.util.List;
 import java.util.Objects;
@@ -81,8 +86,16 @@ public class PlayerChunkSender {
     public static void sendChunk(ServerGamePacketListenerImpl handler, ServerLevel world, LevelChunk chunk) { // Paper - public
         // Paper start - Anti-Xray
         final boolean shouldModify = world.chunkPacketBlockController.shouldModify(handler.player, chunk);
-        handler.send(new ClientboundLevelChunkWithLightPacket(chunk, world.getLightEngine(), null, null, shouldModify));
-        // Paper end - Anti-Xray
+        // ShreddedPaper start - Cache chunk packet
+        Reference<ClientboundLevelChunkWithLightPacket> cachedChunkPacket = chunk.cachedChunkPacket;
+        ClientboundLevelChunkWithLightPacket chunkPacket;
+        if (cachedChunkPacket == null || (chunkPacket = chunk.cachedChunkPacket.get()) == null || chunkPacket.modifyBlocks != shouldModify) {
+            chunkPacket = new ClientboundLevelChunkWithLightPacket(chunk, world.getLightEngine(), null, null, shouldModify);
+            if (ShreddedPaperConfiguration.get().optimizations.chunkPacketCaching.enabled) chunk.cachedChunkPacket = ShreddedPaperConfiguration.get().optimizations.chunkPacketCaching.useSoftReferences ? new SoftReference<>(chunkPacket) : new WeakReference<>(chunkPacket);
+        }
+        chunk.cachedChunkPacketLastAccessed = world.getGameTime();
+        handler.send(chunkPacket);
+        // ShreddedPaper end - Cache chunk packet        // Paper end - Anti-Xray
         // Paper start - PlayerChunkLoadEvent
         if (io.papermc.paper.event.packet.PlayerChunkLoadEvent.getHandlerList().getRegisteredListeners().length > 0) {
             new io.papermc.paper.event.packet.PlayerChunkLoadEvent(new org.bukkit.craftbukkit.CraftChunk(chunk), handler.getPlayer().getBukkitEntity()).callEvent();
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index a73df22ca0a021383df3734aa338adb5bda2084b..6da5f7fae60821e41c5754169a5cce86548d99aa 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -700,7 +700,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
     public final void discard() {
         // CraftBukkit start - add Bukkit remove cause
-        this.discard(null);
+        discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // ShreddedPaper - add Bukkit remove cause;
     }
 
     public final void discard(EntityRemoveEvent.Cause cause) {
diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
index e2720ecc6bab0deb3786f3860dfe4948af86c661..4b42dc74209d8a4b6bcb4e1ba40754334d5a6264 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
@@ -7,6 +7,9 @@ import com.mojang.logging.LogUtils;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
 import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import it.unimi.dsi.fastutil.shorts.ShortListIterator;
+
+import java.lang.ref.Reference;
+import java.lang.ref.WeakReference;
 import java.util.Iterator;
 import java.util.Map;
 import java.util.Map.Entry;
@@ -23,6 +26,7 @@ import net.minecraft.core.registries.Registries;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.game.ClientboundLevelChunkPacketData;
+import net.minecraft.network.protocol.game.ClientboundLevelChunkWithLightPacket;
 import net.minecraft.server.level.FullChunkStatus;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.util.profiling.ProfilerFiller;
@@ -86,6 +90,9 @@ public class LevelChunk extends ChunkAccess implements ca.spottedleaf.moonrise.p
     private final LevelChunkTicks<Block> blockTicks;
     private final LevelChunkTicks<Fluid> fluidTicks;
 
+    public @Nullable Reference<ClientboundLevelChunkWithLightPacket> cachedChunkPacket; // ShreddedPaper
+    public long cachedChunkPacketLastAccessed; // ShreddedPaper
+
     // Pufferfish start - instead of using a random every time the chunk is ticked, define when lightning strikes preemptively
     private int lightningTick;
     // shouldDoLightning compiles down to 29 bytes, which with the default of 35 byte inlining should guarantee an inline
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 5c4dfae5628fc31c72a212cba351d0d36f6cb3dc..0f818128df21cf2532e6428c67975abb8c008de1 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -6,6 +6,7 @@ import com.google.common.collect.ImmutableSet;
 import com.google.common.io.BaseEncoding;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.util.Pair;
+import io.multipaper.shreddedpaper.config.ShreddedPaperConfiguration;
 import io.netty.buffer.Unpooled;
 import it.unimi.dsi.fastutil.shorts.ShortArraySet;
 import it.unimi.dsi.fastutil.shorts.ShortSet;
@@ -2252,15 +2253,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     @Override
     public boolean canSee(Player player) {
+        if (ShreddedPaperConfiguration.get().optimizations.disableVanishApi) return true; // ShreddedPaper - optimize canSee
         return this.canSee((org.bukkit.entity.Entity) player);
     }
 
     @Override
     public boolean canSee(org.bukkit.entity.Entity entity) {
+        if (ShreddedPaperConfiguration.get().optimizations.disableVanishApi) return true; // ShreddedPaper - optimize canSee
         return this.equals(entity) || entity.isVisibleByDefault() ^ this.invertedVisibilityEntities.containsKey(entity.getUniqueId()); // SPIGOT-7312: Can always see self
     }
 
     public boolean canSeePlayer(UUID uuid) {
+        if (ShreddedPaperConfiguration.get().optimizations.disableVanishApi) return true; // ShreddedPaper - optimize canSee
         org.bukkit.entity.Entity entity = this.getServer().getPlayer(uuid);
 
         return (entity != null) ? this.canSee(entity) : false; // If we can't find it, we can't see it
