--- a/net/minecraft/server/waypoints/ServerWaypointManager.java
+++ b/net/minecraft/server/waypoints/ServerWaypointManager.java
@@ -28,7 +_,7 @@
     // Paper end - optimize ServerWaypointManager with locator bar disabled
 
     @Override
-    public void trackWaypoint(final WaypointTransmitter waypoint) {
+    public synchronized void trackWaypoint(final WaypointTransmitter waypoint) { // ShreddedPaper - synchronized
         this.waypoints.add(waypoint);
 
         if (!this.locatorBarEnabled) return; // Paper - optimize ServerWaypointManager with locator bar disabled
@@ -38,7 +_,7 @@
     }
 
     @Override
-    public void updateWaypoint(final WaypointTransmitter waypoint) {
+    public synchronized void updateWaypoint(final WaypointTransmitter waypoint) { // ShreddedPaper - synchronized
         if (!this.locatorBarEnabled) return; // Paper - optimize ServerWaypointManager with locator bar disabled
         if (this.waypoints.contains(waypoint)) {
             Map<ServerPlayer, WaypointTransmitter.Connection> playerConnection = Tables.transpose(this.connections).row(waypoint);
@@ -55,13 +_,13 @@
     }
 
     @Override
-    public void untrackWaypoint(final WaypointTransmitter waypoint) {
+    public synchronized void untrackWaypoint(final WaypointTransmitter waypoint) { // ShreddedPaper - synchronized
         this.connections.column(waypoint).forEach((player, connection) -> connection.disconnect());
         Tables.transpose(this.connections).row(waypoint).clear();
         this.waypoints.remove(waypoint);
     }
 
-    public void addPlayer(final ServerPlayer player) {
+    public synchronized void addPlayer(final ServerPlayer player) { // ShreddedPaper - synchronized
         this.players.add(player);
 
         if (this.locatorBarEnabled) { // Paper - optimize ServerWaypointManager with locator bar disabled
@@ -75,7 +_,7 @@
         }
     }
 
-    public void updatePlayer(final ServerPlayer player) {
+    public synchronized void updatePlayer(final ServerPlayer player) { // ShreddedPaper - synchronized
         if (!this.locatorBarEnabled) return; // Paper - optimize ServerWaypointManager with locator bar disabled
         Map<WaypointTransmitter, WaypointTransmitter.Connection> waypointConnections = this.connections.row(player);
         SetView<WaypointTransmitter> potentialWaypoints = Sets.difference(this.waypoints, waypointConnections.keySet());
@@ -89,7 +_,7 @@
         }
     }
 
-    public void removePlayer(final ServerPlayer player) {
+    public synchronized void removePlayer(final ServerPlayer player) { // ShreddedPaper - synchronized
         this.connections.row(player).values().removeIf(connection -> {
             connection.disconnect();
             return true;
@@ -98,20 +_,20 @@
         this.players.remove(player);
     }
 
-    public void breakAllConnections() {
+    public synchronized void breakAllConnections() { // ShreddedPaper - synchronized
         this.connections.values().forEach(WaypointTransmitter.Connection::disconnect);
         this.connections.clear();
     }
 
-    public void remakeConnections(final WaypointTransmitter waypoint) {
+    public synchronized void remakeConnections(final WaypointTransmitter waypoint) { // ShreddedPaper - synchronized
         if (!this.locatorBarEnabled) return; // Paper - optimize ServerWaypointManager with locator bar disabled
         for (ServerPlayer player : this.players) {
             this.createConnection(player, waypoint);
         }
     }
 
-    public Set<WaypointTransmitter> transmitters() {
-        return this.waypoints;
+    public synchronized Set<WaypointTransmitter> transmitters() { // ShreddedPaper - synchronized
+        return Set.copyOf(this.waypoints); // ShreddedPaper - synchronized - make a copy
     }
 
     // Paper start - optimize ServerWaypointManager with locator bar disabled
